// @comunica/actor-http-fetch/lib/FetchInitPreprocessor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FetchInitPreprocessor = void 0;
/* eslint-disable import/no-nodejs-modules */
const node_http_1 = require("node:http");
const node_https_1 = require("node:https");
const context_entries_1 = require("@comunica/context-entries");
const undici_1 = require("undici");
const MetadataWrappingCacheStore_1 = require("./MetadataWrappingCacheStore");
// eslint-disable-next-line ts/no-require-imports,ts/no-var-requires
const cacheStores = require('undici').cacheStores;
// TODO: the require above can be replaced with import when undici updates.
/**
 * Overrides the HTTP agent to perform better in Node.js.
 */
class FetchInitPreprocessor {
    agent;
    dispatcher;
    cache;
    constructor(args) {
        const httpAgent = new node_http_1.Agent(args.agentOptions);
        const httpsAgent = new node_https_1.Agent(args.agentOptions);
        this.agent = (_parsedURL) => _parsedURL.protocol === 'http:' ? httpAgent : httpsAgent;
        this.createCache(args);
        args.httpInvalidator.addInvalidateListener(({ url }) => {
            if (url) {
                const parsedUrl = new URL(url);
                // eslint-disable-next-line ts/no-floating-promises
                this.cache.delete({ origin: parsedUrl.origin, path: parsedUrl.pathname + parsedUrl.search, method: 'GET' });
            }
            else {
                this.createCache(args);
            }
        });
    }
    createCache(args) {
        this.cache = new cacheStores.MemoryCacheStore({
            maxSize: args.cacheMaxSize,
            maxCount: args.cacheMaxCount,
            maxEntrySize: args.cacheMaxEntrySize,
        });
        this.dispatcher = new undici_1.Agent().compose(undici_1.interceptors.cache({
            store: new MetadataWrappingCacheStore_1.MetadataWrappingCacheStore(this.cache),
            methods: ['GET', 'HEAD'],
        }));
    }
    async handle(init, context) {
        // Add 'Accept-Encoding' headers
        const headers = new Headers(init.headers);
        if (!headers.has('Accept-Encoding')) {
            headers.set('Accept-Encoding', 'br,gzip,deflate');
            init = { ...init, headers };
        }
        // The Fetch API requires specific options to be set when sending body streams:
        // - 'keepalive' can not be true
        // - 'duplex' must be set to 'half'
        return {
            ...init,
            ...init.body ? { keepalive: false, duplex: 'half' } : { keepalive: true },
            agent: this.agent,
            dispatcher: context.get(context_entries_1.KeysHttp.httpCache) ? this.dispatcher : undefined,
        };
    }
}
exports.FetchInitPreprocessor = FetchInitPreprocessor;
