// @rdfc/sds-processors-ts/lib/utils/index.js
import { DataFactory } from "rdf-data-factory";
import { Parser } from "n3";
import { SDS, XSD } from "@treecg/types";
import { extractShapes, match, subject, } from "rdf-lens";
import { RdfStore } from "rdf-stores";
const df = new DataFactory();
export const SDS_GRAPH = SDS.terms.custom("DataDescription");
export const SHAPES_TEXT = "@prefix tree: <https://w3id.org/tree#>.\n@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.\n@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.\n@prefix sds: <https://w3id.org/sds#>.\n@prefix sh: <http://www.w3.org/ns/shacl#>.\n@prefix xsd: <http://www.w3.org/2001/XMLSchema#>.\n@prefix rdfl: <https://w3id.org/rdf-lens/ontology#>.\n\n[ ] a sh:NodeShape;\n  sh:targetClass <RdfThing>;\n  sh:property [\n    sh:name \"id\";\n    sh:path ( );\n    sh:maxCount 1;\n    sh:minCount 1;\n    sh:datatype xsd:iri;\n  ], [\n    sh:name \"quads\";\n    sh:path ( );\n    sh:maxCount 1;\n    sh:minCount 1;\n    sh:class rdfl:CBD;\n  ].\n\n[ ] a sh:NodeShape;\n  sh:targetClass <#Bucket>;\n  sh:property [\n    sh:name \"links\";\n    sh:path sds:relation;\n    sh:class <#Relation>;\n  ], [\n    sh:name \"id\";\n    sh:path ( );\n    sh:datatype xsd:iri;\n    sh:minCount 1;\n    sh:maxCount 1;\n  ], [\n    sh:name \"parent\";\n    sh:path [ sh:inversePath ( sds:relation sds:relationBucket ) ];\n    sh:class <#Bucket>;\n    sh:maxCount 1;\n  ], [\n    sh:name \"root\";\n    sh:path sds:isRoot;\n    sh:datatype xsd:boolean;\n    sh:maxCount 1;\n  ], [\n    sh:name \"immutable\";\n    sh:path sds:immutable;\n    sh:datatype xsd:boolean;\n    sh:maxCount 1;\n  ].\n\n[ ] a sh:NodeShape;\n  sh:targetClass <#Relation>;\n  sh:property [\n    sh:name \"type\";\n    sh:path sds:relationType;\n    sh:datatype xsd:iri;\n    sh:minCount 1;\n    sh:maxCount 1;\n  ], [\n    sh:name \"target\";\n    sh:path sds:relationBucket;\n    sh:datatype xsd:iri;\n    sh:minCount 1;\n    sh:maxCount 1;\n  ], [\n    sh:name \"value\";\n    sh:path sds:relationValue;\n    sh:datatype xsd:literal;\n    sh:maxCount 1;\n  ], [\n    sh:name \"path\";\n    sh:path sds:relationPath;\n    sh:class <RdfThing>;\n    sh:maxCount 1;\n  ].\n\n[ ] a sh:NodeShape;\n  sh:targetClass <#Record>;\n  sh:property [\n    sh:name \"stream\";\n    sh:path sds:stream;\n    sh:datatype xsd:iri;\n    sh:minCount 1;\n    sh:maxCount 1;\n  ], [\n    sh:name \"data\";\n    sh:path sds:payload;\n    sh:datatype xsd:iri;\n    sh:minCount 1;\n    sh:maxCount 1;\n  ], [\n    sh:name \"bucket\";\n    sh:path sds:bucket;\n    sh:class <#Bucket>;\n    sh:maxCount 1;\n  ].\n\n";
function writeRelation(rel, writer) {
    const id = df.blankNode();
    writer.addQuads([
        df.quad(id, SDS.terms.relationType, rel.type, SDS.terms.custom("DataDescription")),
        df.quad(id, SDS.terms.relationBucket, rel.target, SDS.terms.custom("DataDescription")),
    ]);
    if (rel.value) {
        writer.addQuad(df.quad(id, SDS.terms.relationValue, rel.value, SDS.terms.custom("DataDescription")));
    }
    if (rel.path) {
        writer.addQuads([
            df.quad(id, SDS.terms.relationPath, rel.path.id, SDS.terms.custom("DataDescription")),
            ...rel.path.quads.map((x) => df.quad(x.subject, x.predicate, x.object, SDS.terms.custom("DataDescription"))),
        ]);
    }
    return id;
}
export class Bucket {
    id;
    parent;
    root;
    immutable;
    links;
    empty = false;
    addMember;
    constructor(id, links, root, immutable, parent) {
        this.id = id;
        this.root = root;
        this.immutable = immutable;
        this.links = links;
        this.parent = parent;
    }
    static parse(bucket, bucket_cache) {
        const parent = bucket.parent
            ? Bucket.parse(bucket.parent, bucket_cache)
            : undefined;
        const links = bucket.links.map(({ target, path, value, type }) => ({
            path,
            value,
            type,
            target: target,
        }));
        const out = new Bucket(bucket.id, links, bucket.root, bucket.immutable, parent);
        bucket_cache[bucket.id.value] = out;
        return out;
    }
    write(writer) {
        const id = this.id;
        const relations = this.links
            .map((rel) => writeRelation(rel, writer))
            .map((rel) => df.quad(id, SDS.terms.relation, rel, SDS.terms.custom("DataDescription")));
        if (this.root) {
            relations.push(df.quad(id, SDS.terms.custom("isRoot"), df.literal("true", XSD.terms.custom("boolean")), SDS.terms.custom("DataDescription")));
        }
        writer.addQuads(relations);
    }
}
export class Record {
    stream;
    data;
    bucket;
    constructor(data, stream, bucket) {
        this.stream = stream;
        this.data = data;
        this.bucket = bucket;
    }
    static parse({ stream, data, bucket }, store, bucket_cache) {
        const quads = store
            .getQuads()
            .filter((q) => q.graph.value !== SDS.custom("DataDescription"));
        let actual_bucket;
        if (bucket) {
            actual_bucket = Bucket.parse(bucket, bucket_cache);
        }
        return new Record({ id: data, quads }, stream, actual_bucket);
    }
    write(writer) {
        const id = df.blankNode();
        const quads = [
            df.quad(id, SDS.terms.payload, this.data.id, SDS.terms.custom("DataDescription")),
            df.quad(id, SDS.terms.stream, this.stream, SDS.terms.custom("DataDescription")),
        ];
        if (this.bucket) {
            quads.push(df.quad(id, SDS.terms.bucket, this.bucket.id, SDS.terms.custom("DataDescription")));
            this.bucket.write(writer);
        }
        writer.addQuads(quads);
    }
}
export class Extractor {
    extractor;
    shapes;
    lens;
    shape;
    bucket_cache = {};
    constructor(extractor, stream) {
        this.extractor = extractor;
        const quads = new Parser({ baseIRI: "" }).parse(SHAPES_TEXT);
        this.shapes = extractShapes(quads, {
            "#Bucket": (item) => {
                return Bucket.parse(item, this.bucket_cache);
            },
        });
        this.lens = match(undefined, SDS.terms.stream, stream)
            .thenAll(subject)
            .thenSome(this.shapes.lenses["#Record"])
            .mapAll((x) => x);
    }
    async parse_records(quads) {
        const store = RdfStore.createDefault();
        quads.forEach((quad) => store.addQuad(quad));
        const dtos = this.lens.execute(quads);
        return await Promise.all(dtos.map((dto) => {
            return Record.parse(dto, store, this.bucket_cache);
        }));
    }
}
export async function getSubjects(store, pred, object, graph) {
    const quads = await store.match(null, pred, object, graph).toArray();
    return quads.map((x) => x.subject);
}
export async function getObjects(store, subject, pred, graph) {
    const quads = await store.match(subject, pred, null, graph).toArray();
    return quads.map((x) => x.object);
}
export function getOrDefaultMap(map, key, def) {
    return map.get(key) || map.set(key, def).get(key);
}
export function maybeParse(data) {
    if (typeof data === "string" || data instanceof String) {
        const parse = new Parser();
        return parse.parse(data);
    }
    else {
        return data;
    }
}
export function findAndValidateRecordTimestamp(record, path, lastMemberTimestamp, logger) {
    let recordTimestamp = undefined;
    if (path) {
        const values = path
            .execute(record.data)
            .filter((x, i, arr) => arr.findIndex((y) => x.value === y.value) == i)
            .filter((value) => value.literal);
        if (values.length !== 1) {
            logger.error(`Expected exactly one timestamp value, got ${values.length}. Ignoring record '${record.data.id.value}'.`);
            return null;
        }
        recordTimestamp = new Date(values[0].literal.value);
        if (recordTimestamp.getTime() < lastMemberTimestamp) {
            logger.error(`Record timestamp is before the last record timestamp. Are your records out of order? Ignoring record '${record.data.id.value}'.`);
            return null;
        }
    }
    return recordTimestamp;
}
