// @rdfc/dumps-to-feed-processor-ts/dist/index.js
import { rdfDereferencer } from "rdf-dereference";
import { CBDShapeExtractor } from "extract-cbd-shape";
import { createHash } from 'node:crypto';
import { RdfStore } from 'rdf-stores';
import { DataFactory } from 'rdf-data-factory';
import { Writer as NWriter } from 'n3';
import { Level } from "level";
import { QueryEngine } from "@comunica/query-sparql";
import { rdfParser } from "rdf-parse";
import { arrayifyStream } from "arrayify-stream";
import streamifyString from "streamify-string";
import { streamifyArray } from "streamify-array";
import * as path from "path";
import { getLoggerFor } from "./utils/logUtil.js";
import Queue from "queue-fifo";
import { Readable } from 'stream';
import { canonize } from 'rdf-canonize';
const logger = getLoggerFor("processor");
const df = new DataFactory();
const engine = new QueryEngine();
async function loadQuadStreamInStore(store, quadStream) {
    return new Promise((resolve, reject) => {
        store.import(quadStream).on("end", resolve).on("error", reject);
    });
}
async function processActivity(writer, quads, type, iri, hash) {
    let nWriter = new NWriter({ "format": "application/trig" });
    let subject = df.namedNode("#" + hash);
    nWriter.addQuads([
        df.quad(subject, df.namedNode("http://www.w3.org/1999/02/22-rdf-syntax-ns#type"), type),
        df.quad(subject, df.namedNode("https://www.w3.org/ns/activitystreams#object"), iri),
        df.quad(subject, df.namedNode("https://www.w3.org/ns/activitystreams#published"), df.literal((new Date()).toISOString(), df.namedNode('http://www.w3.org/2001/XMLSchema#dateTime'))),
    ]);
    for (let quad of quads) {
        nWriter.addQuad(quad.subject, quad.predicate, quad.object, subject);
    }
    return new Promise((resolve, reject) => {
        nWriter.end(async (error, result) => {
            await writer.push(result);
            resolve(undefined);
        });
    });
}
async function dumpToRdfStore(dump, dumpContentType) {
    const store = RdfStore.createDefault();
    if (Buffer.isBuffer(dump)) {
        for (const quad of await arrayifyStream(rdfParser.parse(Readable.from(dump), { contentType: dumpContentType }))) {
            store.addQuad(quad);
        }
    }
    else if (dumpContentType === 'identifier') {
        const { data } = await rdfDereferencer.dereference(dump, { localFiles: true });
        await loadQuadStreamInStore(store, data);
    }
    else {
        for (const quad of await arrayifyStream(rdfParser.parse(streamifyString(dump), { contentType: dumpContentType }))) {
            store.addQuad(quad);
        }
    }
    return store;
}
async function focusNodesToSubjects(store, focusNodesStrategy, focusNodes) {
    switch (focusNodesStrategy) {
        case 'extract':
            return findFocusNodes(store);
        case 'sparql':
            return findFocusNodes(store, focusNodes);
        case 'iris':
            if (!focusNodes) {
                logger.error(`[focusNodesToSubjects] focusNodesStrategy is set to '${focusNodesStrategy}' but no focusNodes were provided`);
                return streamifyArray([]);
            }
            return streamifyArray(focusNodes.split(',').map((iri) => df.namedNode(iri)));
    }
}
async function findFocusNodes(store, query) {
    query = query || `
  PREFIX dcat: <http://www.w3.org/ns/dcat#>
  PREFIX foaf: <http://xmlns.com/foaf/0.1/>
  PREFIX vcard: <http://www.w3.org/2006/vcard/ns#>
  PREFIX dcterms: <http://purl.org/dc/terms/>

  SELECT DISTINCT ?entity
  WHERE {
    ?entity a ?type .
    FILTER (?type IN (dcat:Catalog, dcat:Dataset, dcat:Distribution, dcat:DataService, foaf:Agent, vcard:Kind, dcterms:LicenseDocument))
  }
  `;
    const bindingsStream = await engine.queryBindings(query, { sources: [store] });
    return (bindingsStream.map((bindings) => bindings.get('entity')).filter((term) => term !== undefined));
}
export async function main(writer, feedname, flush, dump, dumpContentType, focusNodesStrategy, nodeShapeIri, nodeShape, focusNodes, dbDir = "./") {
    const db = new Level(path.join(dbDir, "state-of-" + feedname), { valueEncoding: 'json' });
    if (flush) {
        logger.info(`Flushing the database for feed ${feedname}.`);
        await db.clear();
    }
    const store = await dumpToRdfStore(dump, dumpContentType);
    let nodeShapeStore;
    if (nodeShape) {
        nodeShapeStore = await dumpToRdfStore(nodeShape, 'text/turtle');
    }
    const extractor = new CBDShapeExtractor(nodeShapeStore);
    const subjectsStream = await focusNodesToSubjects(store, focusNodesStrategy, focusNodes);
    const nodeShapeIriTerm = df.namedNode(nodeShapeIri);
    let processing = 0;
    const subjectsQueue = new Queue();
    let busy = false;
    subjectsStream.on('data', async (subjectToQueue) => {
        subjectsQueue.enqueue(subjectToQueue);
        if (busy) {
            return;
        }
        busy = true;
        while (!subjectsQueue.isEmpty()) {
            const subject = subjectsQueue.dequeue();
            processing++;
            if (subject.termType === 'BlankNode') {
                logger.error("An entity (" + subject.value + ") cannot be a blank node!");
                processing--;
            }
            else if (subject.termType === 'NamedNode') {
                let entityQuads = await extractor.extract(store, subject, nodeShapeIriTerm);
                let canonizedString = await canonize(entityQuads, { algorithm: 'RDFC-1.0' });
                let hashString = createHash('md5').update(canonizedString).digest('hex');
                try {
                    let previousHashString = await db.get(subject.value);
                    if (previousHashString !== hashString) {
                        logger.debug(`Unequal hashes found in the db for '${subject.value}'. Processing an Update.`);
                        await db.put(subject.value, hashString);
                        await processActivity(writer, entityQuads, df.namedNode("https://www.w3.org/ns/activitystreams#Update"), subject, hashString);
                    }
                    else {
                        logger.verbose(`Equal hashes found in the db for '${subject.value}'. Doing nothing.`);
                    }
                }
                catch (e) {
                    logger.debug(`No hash found in the db for '${subject.value}'. Processing a Create.`);
                    await db.put(subject.value, hashString);
                    await processActivity(writer, entityQuads, df.namedNode("https://www.w3.org/ns/activitystreams#Create"), subject, hashString);
                }
                finally {
                    processing--;
                }
            }
        }
        busy = false;
    });
    await new Promise((resolve, reject) => {
        subjectsStream.on('end', async () => {
            while (processing > 0) {
                logger.verbose("Waiting for processing to finish...");
                await new Promise((resolve) => setTimeout(resolve, 100));
            }
            let keys = await db.keys().all();
            for (let key of keys) {
                if (store.getQuads(df.namedNode(key), null, null).length === 0) {
                    logger.debug(`Key '${key}' was found in the db but not in the store. Processing a Delete.`);
                    await processActivity(writer, [], df.namedNode("https://www.w3.org/ns/activitystreams#Delete"), df.namedNode(key), "deletion-" + encodeURIComponent(new Date().toISOString()));
                    await db.del(key);
                }
            }
            await db.close();
            logger.debug("Finished processing the stream. Closing the leveldb.");
            resolve(undefined);
        });
        subjectsStream.on('error', (e) => {
            logger.error(e);
            reject(e);
        });
    });
}
export async function processor(writer, feedname, flush, dump, dumpContentType, focusNodesStrategy, nodeShapeIri, nodeShape, focusNodes, dbDir = "./") {
    const listenToNodeShape = !!nodeShape;
    const listenToFocusNodes = focusNodesStrategy === 'iris' || focusNodesStrategy === 'sparql';
    if (listenToFocusNodes && !focusNodes) {
        logger.error(`focusNodesStrategy is set to '${focusNodesStrategy}' but no focusNodes were provided`);
        return;
    }
    const dumpBuffer = [];
    const nodeShapeBuffer = [];
    const focusNodesBuffer = [];
    let tryProcessing = async () => {
        while (dumpBuffer.length && (nodeShapeBuffer.length || !listenToNodeShape) && (focusNodesBuffer.length || !listenToFocusNodes)) {
            const nextDump = dumpBuffer.shift();
            const nextNodeShape = nodeShapeBuffer.shift();
            const nextFocusNodes = focusNodesBuffer.shift();
            await main(writer, feedname, flush, nextDump, dumpContentType, focusNodesStrategy, nodeShapeIri, nextNodeShape, nextFocusNodes, dbDir);
        }
    };
    dump.data(async (data) => {
        dumpBuffer.push(data);
        await tryProcessing();
    });
    if (listenToNodeShape) {
        nodeShape.data(async (data) => {
            nodeShapeBuffer.push(data);
            await tryProcessing();
        });
    }
    if (listenToFocusNodes) {
        focusNodes.data(async (data) => {
            focusNodesBuffer.push(data);
            await tryProcessing();
        });
    }
    let dumpEnded = false;
    let nodeShapeEnded = false;
    let focusNodesEnded = false;
    let tryClosingWriter = () => {
        if (dumpEnded && nodeShapeEnded && focusNodesEnded) {
            logger.info("All input streams ended, closing writer.");
            writer.end();
        }
    };
    dump.on('end', async () => {
        dumpEnded = true;
        tryClosingWriter();
    });
    if (listenToNodeShape) {
        nodeShape.on('end', async () => {
            nodeShapeEnded = true;
            tryClosingWriter();
        });
    }
    else {
        nodeShapeEnded = true;
    }
    if (listenToFocusNodes) {
        focusNodes.on('end', async () => {
            focusNodesEnded = true;
            tryClosingWriter();
        });
    }
    else {
        focusNodesEnded = true;
    }
}
