// @rdfc/js-runner/dist/connectors/http.js
import * as http from "http";
import { createServer } from "http";
import { SimpleStream, } from "../connectors.js";
function streamToString(stream, binary) {
    const datas = [];
    return new Promise((res) => {
        stream.on("data", (data) => {
            datas.push(data);
        });
        stream.on("end", () => {
            const streamData = Buffer.concat(datas);
            res(binary ? streamData : streamData.toString());
        });
    });
}
export const startHttpStreamReader = (config) => {
    let server = undefined;
    const stream = new SimpleStream(() => new Promise((res) => {
        if (server !== undefined) {
            server.close(() => {
                res();
            });
        }
        else {
            res();
        }
    }));
    const requestListener = async function (req, res) {
        try {
            const content = await streamToString(req, config.binary);
            const promise = stream.push(content).catch((error) => {
                throw error;
            });
            if (config.waitHandled) {
                await promise;
            }
        }
        catch (error) {
            console.error("Failed", error);
        }
        res.writeHead(config.responseCode || 200);
        res.end("OK");
    };
    server = createServer(requestListener);
    const init = () => {
        return new Promise((res) => {
            const cb = () => res(undefined);
            if (server) {
                server.listen(config.port, config.endpoint, cb);
            }
            else {
                cb();
            }
        });
    };
    return { reader: stream, init };
};
export const startHttpStreamWriter = (config) => {
    const requestConfig = new URL(config.endpoint);
    const writer = new SimpleStream();
    writer.push = async (item) => {
        await new Promise((resolve) => {
            const options = {
                hostname: requestConfig.hostname,
                path: requestConfig.path,
                method: config.method,
                port: requestConfig.port,
            };
            const cb = (response) => {
                response.on("data", () => { });
                response.on("end", () => {
                    resolve(null);
                });
            };
            const req = http.request(options, cb);
            req.write(item, () => {
                req.end();
            });
        });
    };
    return { writer, init: async () => { } };
};
