// @traqula/algebra-transformations-1-2/dist/cjs/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.algebraUtils = exports.Canonicalizer = exports.ExpressionTypes = exports.Types = exports.AlgebraFactory = exports.Algebra = void 0;
const algebra_transformations_1_1_1 = require("@traqula/algebra-transformations-1-1");
const algebraUtils12 = __importStar(require("./utils.js"));
__exportStar(require("./toAlgebra12.js"), exports);
__exportStar(require("./toAst12.js"), exports);
__exportStar(require("./types.js"), exports);
var algebra_transformations_1_1_2 = require("@traqula/algebra-transformations-1-1");
Object.defineProperty(exports, "Algebra", { enumerable: true, get: function () { return algebra_transformations_1_1_2.Algebra; } });
Object.defineProperty(exports, "AlgebraFactory", { enumerable: true, get: function () { return algebra_transformations_1_1_2.AlgebraFactory; } });
Object.defineProperty(exports, "Types", { enumerable: true, get: function () { return algebra_transformations_1_1_2.Types; } });
Object.defineProperty(exports, "ExpressionTypes", { enumerable: true, get: function () { return algebra_transformations_1_1_2.ExpressionTypes; } });
Object.defineProperty(exports, "Canonicalizer", { enumerable: true, get: function () { return algebra_transformations_1_1_2.Canonicalizer; } });
exports.algebraUtils = {
    ...algebra_transformations_1_1_1.algebraUtils,
    ...algebraUtils12,
};
