// @traqula/core/dist/cjs/lib/AstCoreFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AstCoreFactory = void 0;
class AstCoreFactory {
    /**
     * Whether this AstFactory will track source location. Default: true.
     * In case source location is not tracked,
     * each generated node using this factory will be {@link SourceLocationNodeAutoGenerate}
     */
    tracksSourceLocation;
    constructor(args = {}) {
        this.tracksSourceLocation = args.tracksSourceLocation ?? true;
    }
    /**
     * Wrap any type into an object that tracks the source location of tha value.
     * @param val the value to wrap
     * @param loc the source location for that value
     */
    wrap(val, loc) {
        return { val, loc };
    }
    /**
     * Whether the provided value is an object that tracks source location.
     * @param obj
     */
    isLocalized(obj) {
        return typeof obj === 'object' && obj !== null && 'loc' in obj &&
            typeof obj.loc === 'object' && obj.loc !== null && 'sourceLocationType' in obj.loc;
    }
    /**
     * Compute the source location of an element based on the elements it contains.
     * The provided arguments should be in order of occurrence in the string.
     */
    sourceLocation(...elements) {
        if (!this.tracksSourceLocation) {
            return this.gen();
        }
        const pureElements = elements.filter(x => x !== undefined);
        if (pureElements.length === 0) {
            return this.sourceLocationNoMaterialize();
        }
        const filtered = pureElements.filter(element => !this.isLocalized(element) || this.isSourceLocationSource(element.loc) ||
            this.isSourceLocationStringReplace(element.loc) || this.isSourceLocationNodeReplace(element.loc));
        if (filtered.length === 0) {
            return this.gen();
        }
        const first = filtered.at(0);
        const last = filtered.at(-1);
        return {
            sourceLocationType: 'source',
            start: this.isLocalized(first) ?
                first.loc.start :
                first.startOffset,
            end: this.isLocalized(last) ?
                last.loc.end :
                (last.endOffset + 1),
        };
    }
    /**
     * Generate a source location indicating the no materialization.
     */
    sourceLocationNoMaterialize() {
        if (!this.tracksSourceLocation) {
            return this.gen();
        }
        return { sourceLocationType: 'noMaterialize' };
    }
    /**
     * Returns a copy of the argument that is not materialized
     */
    dematerialized(arg) {
        return { ...arg, loc: this.sourceLocationNoMaterialize() };
    }
    /**
     * Given a value and some mapper from objects to objects,
     * map all containing items in case it is an array,
     * otherwise map the provided value in case it is an object, finally, do nothing.
     * @param value the value to map
     * @param mapper the function mapping an object to another object.
     */
    safeObjectTransform(value, mapper) {
        if (value && typeof value === 'object') {
            // If you wonder why this is all so hard, this is the reason. We cannot lose the methods of our Array objects
            if (Array.isArray(value)) {
                return value.map(x => this.safeObjectTransform(x, mapper));
            }
            return mapper(value);
        }
        return value;
    }
    /**
     * Given a (AST) tree, return a copy of that tree where it and it's
     * descendents all have a {@link SourceLocationNodeAutoGenerate} localization.
     */
    forcedAutoGenTree(obj) {
        const copy = { ...obj };
        for (const [key, value] of Object.entries(copy)) {
            copy[key] = this.safeObjectTransform(value, obj => this.forcedAutoGenTree(obj));
        }
        if (this.isLocalized(copy)) {
            copy.loc = this.gen();
        }
        return copy;
    }
    /**
     * In case the provided Node is not yet materialized, force it to be autoGenerated.
     */
    forceMaterialized(arg) {
        if (this.isSourceLocationNoMaterialize(arg.loc)) {
            return this.forcedAutoGenTree(arg);
        }
        return { ...arg };
    }
    /**
     * Check whether an object is in fact a {@link SourceLocation}.
     * @param loc
     */
    isSourceLocation(loc) {
        return 'sourceLocationType' in loc;
    }
    /**
     * Create a {@link SourceLocation} that indicates what range of characters this node represents in the source string.
     */
    sourceLocationSource(start, end) {
        return {
            sourceLocationType: 'source',
            start,
            end,
        };
    }
    /**
     * Create a {@link SourceLocation} that indicates a given range of the current range should be replaced by
     * a new source.
     */
    sourceLocationInlinedSource(newSource, subLoc, start, end, startOnNew = 0, endOnNew = newSource.length) {
        if (!this.tracksSourceLocation) {
            return this.gen();
        }
        if (this.isSourceLocationSource(subLoc)) {
            startOnNew = subLoc.start;
            endOnNew = subLoc.end;
        }
        return {
            sourceLocationType: 'inlinedSource',
            newSource,
            start,
            end,
            loc: subLoc,
            startOnNew,
            endOnNew,
        };
    }
    ;
    /**
     * Guard to check if an object is a {@link SourceLocation} of type {@link SourceLocationInlinedSource}.
     */
    isSourceLocationInlinedSource(loc) {
        return this.isSourceLocation(loc) && loc.sourceLocationType === 'inlinedSource';
    }
    /**
     * Create a {@link SourceLocation} indicating the node should be autoGenerated by a generator.
     */
    gen() {
        return { sourceLocationType: 'autoGenerate' };
    }
    /**
     * Guard to check if an object is a {@link SourceLocation} of type {@link SourceLocationSource}.
     */
    isSourceLocationSource(loc) {
        return this.isSourceLocation(loc) && loc.sourceLocationType === 'source';
    }
    /**
     * Create a {@link SourceLocation} that indicates this node,
     * representing a range of characters in the original string, should be replaced by some string during generation.
     */
    sourceLocationStringReplace(newSource, start, end) {
        if (!this.tracksSourceLocation) {
            return this.gen();
        }
        return { sourceLocationType: 'stringReplace', newSource, start, end };
    }
    /**
     * Guard to check if an object is a {@link SourceLocation} of type {@link SourceLocationStringReplace}.
     * @param loc
     */
    isSourceLocationStringReplace(loc) {
        return this.isSourceLocation(loc) && loc.sourceLocationType === 'stringReplace';
    }
    /**
     * Given a sourceLocation, generate a new {@link SourceLocation}
     * that indicates the range of the given location should now be autoGenerated
     */
    sourceLocationNodeReplaceUnsafe(loc) {
        if (this.isSourceLocationSource(loc)) {
            return this.sourceLocationNodeReplace(loc);
        }
        if (this.isSourceLocationInlinedSource(loc)) {
            return this.sourceLocationNodeReplaceUnsafe(loc.loc);
        }
        throw new Error(`Cannot convert SourceLocation of type ${loc.sourceLocationType} to SourceLocationNodeReplace`);
    }
    sourceLocationNodeReplace(startOrLoc, end) {
        let starting;
        let ending;
        if (typeof startOrLoc === 'number') {
            starting = startOrLoc;
            ending = end;
        }
        else {
            starting = startOrLoc.start;
            ending = startOrLoc.end;
        }
        return {
            sourceLocationType: 'nodeReplace',
            start: starting,
            end: ending,
        };
    }
    /**
     * Guard to check if an object is a {@link SourceLocation} of type {@link SourceLocationNodeReplace}.
     */
    isSourceLocationNodeReplace(loc) {
        return this.isSourceLocation(loc) && loc.sourceLocationType === 'nodeReplace';
    }
    /**
     * Guard to check if an object is a {@link SourceLocation} of type {@link SourceLocationNodeAutoGenerate}.
     */
    isSourceLocationNodeAutoGenerate(loc) {
        return this.isSourceLocation(loc) && loc.sourceLocationType === 'autoGenerate';
    }
    /**
     * Check whether the provided {@link SourceLocation} expects the generator to autoGenerate.
     */
    isPrintingLoc(loc) {
        return this.isSourceLocationNodeReplace(loc) ||
            this.isSourceLocationNodeAutoGenerate(loc) ||
            (this.isSourceLocationInlinedSource(loc) && this.isPrintingLoc(loc.loc));
    }
    /**
     * A simple filter that will only execute the provided callback when the provided {@link SourceLocation}
     * expects the generator steps to autoGenerate.
     */
    printFilter(node, callback) {
        if (this.isPrintingLoc(node.loc)) {
            callback();
        }
    }
    /**
     * Guard to check if an object is a {@link SourceLocation} of type {@link SourceLocationNoMaterialize}.
     */
    isSourceLocationNoMaterialize(loc) {
        return this.isSourceLocation(loc) && loc.sourceLocationType === 'noMaterialize';
    }
    /**
     * Guard to check if an object is the specified Type: {@link Typed}.
     */
    isOfType(obj, type) {
        const casted = obj;
        return casted.type === type;
    }
    /**
     * Guard to check if an object is the specified Type: {@link SubTyped}.
     */
    isOfSubType(obj, type, subType) {
        const temp = obj;
        return temp.type === type && temp.subType === subType;
    }
}
exports.AstCoreFactory = AstCoreFactory;
