// rdf-lens/dist/index.cjs
'use strict';

var types = require('@treecg/types');
var rdfDataFactory = require('rdf-data-factory');

function termToString$1(term) {
    if (term.termType === "NamedNode") {
        return "<" + term.value + ">";
    }
    if (term.termType === "BlankNode") {
        return "_:" + term.value;
    }
    return JSON.stringify(term.value);
}
class LensError extends Error {
    lineage;
    constructor(message, lineage) {
        super(message);
        this.message = message;
        this.lineage = lineage;
    }
}
function createContext() {
    const ctx = {
        stateMap: new Map(),
        lineage: [],
    };
    const clone = () => ({
        clone,
        stateMap: ctx.stateMap,
        lineage: ctx.lineage.slice(),
    });
    return Object.assign(ctx, { clone });
}
class BasicLens {
    _exec;
    index;
    constructor(execute) {
        this._exec = execute;
    }
    named(name, opts, cb) {
        return new BasicLens((c, ctx) => {
            let extras = asList(opts) || [];
            if (cb) {
                extras = [...extras, ...asList(cb(c))];
            }
            ctx.lineage.push({ name, opts: deconstructList(extras) });
            return this.execute(c, ctx);
        });
    }
    asMulti() {
        return (new BasicLensM((c, ctx) => {
            const out = this.execute(c, ctx);
            return out;
        }));
    }
    and(...and) {
        return (new BasicLens((c, ctx) => {
            const a = this.execute(c, ctx);
            const rest = and.map((x) => x.execute(c, ctx));
            return [a, ...rest];
        }));
    }
    orM(...others) {
        return new BasicLensM((c, ctx) => {
            const all = [this, ...others];
            return all.flatMap((x) => {
                try {
                    return [x.execute(c, ctx.clone())];
                }
                catch (ex) {
                    return [];
                }
            });
        });
    }
    or(...others) {
        return new BasicLens((c, ctx) => {
            const errors = [];
            try {
                return this.execute(c, ctx);
            }
            catch (ex) {
                errors.push(ex);
                for (let i = 0; i < others.length; i++) {
                    try {
                        return others[i].execute(c, ctx.clone());
                    }
                    catch (ex) {
                        errors.push(ex);
                    }
                }
            }
            throw errors;
        });
    }
    map(fn) {
        return new BasicLens((c, ctx) => {
            const a = this.execute(c, ctx);
            return fn(a, ctx);
        });
    }
    then(next) {
        return new BasicLens((c, ctx) => {
            const a = this.execute(c, ctx);
            return next.execute(a, ctx);
        });
    }
    execute(container, ctx = createContext()) {
        return this._exec(container, ctx);
    }
}
function deconstructList(x) {
    if (x.length == 1) {
        return x[0];
    }
    return x;
}
function asList(x) {
    if (Array.isArray(x))
        return x;
    return [x];
}
class BasicLensM extends BasicLens {
    named(name, opts, cb) {
        return new BasicLensM((c, ctx) => {
            let extras = asList(opts) || [];
            if (cb) {
                extras = [...extras, ...asList(cb(c))];
            }
            ctx.lineage.push({ name, opts: deconstructList(extras) });
            return this.execute(c, ctx);
        });
    }
    one(def) {
        return new BasicLens((c, ctx) => {
            const qs = this.execute(c, ctx);
            return qs[0] || def;
        });
    }
    expectOne() {
        return new BasicLens((c, ctx) => {
            const qs = this.execute(c, ctx);
            if (qs.length < 1)
                throw new LensError("Expected one, found none", ctx.lineage.slice());
            return qs[0];
        });
    }
    thenAll(next) {
        return new BasicLensM((c, ctx) => {
            const qs = this.execute(c, ctx.clone());
            return qs.map((x) => next.execute(x, ctx.clone()));
        });
    }
    thenSome(next) {
        return new BasicLensM((c, ctx) => {
            const qs = this.execute(c, ctx.clone());
            return qs.flatMap((x) => {
                try {
                    const o = next.execute(x, ctx.clone());
                    return [o];
                }
                catch (ex) {
                    return [];
                }
            });
        });
    }
    thenFlat(next) {
        return new BasicLensM((c, ctx) => {
            const qs = this.execute(c, ctx.clone());
            return qs.flatMap((x) => next.execute(x, ctx.clone()));
        });
    }
    mapAll(fn) {
        return new BasicLensM((c, ctx) => {
            const qs = this.execute(c, ctx);
            return qs.map((x) => fn(x, ctx));
        });
    }
    orAll(...others) {
        return new BasicLensM((c, ctx) => {
            const out = [];
            try {
                out.push(...this.execute(c, ctx.clone()));
            }
            catch (ex) {
            }
            for (let i = 0; i < others.length; i++) {
                try {
                    out.push(...others[i].execute(c, ctx.clone()));
                }
                catch (ex) {
                }
            }
            return out;
        });
    }
    filter(fn) {
        return new BasicLensM((c, ctx) => {
            return this.execute(c, ctx).filter(fn);
        });
    }
    reduce(lens, start) {
        return new BasicLens((c, ctx) => {
            const st = this.and(start).map(([ts, f]) => {
                return ts.reduce((acc, v) => lens.execute([v, acc], ctx), f);
            });
            return st.execute(c, ctx);
        });
    }
}
function pred(pred) {
    return new BasicLensM(({ quads, id }) => {
        const out = quads.filter((q) => q.subject.equals(id) && (!pred || q.predicate.equals(pred)));
        return out.map((q) => ({ quads, id: q.object }));
    }).named("pred", pred && termToString$1(pred));
}
function invPred(pred) {
    return new BasicLensM(({ quads, id }) => {
        const out = quads.filter((q) => q.object.equals(id) && (!pred || q.predicate.equals(pred)));
        return out.map((q) => ({ quads, id: q.subject }));
    }).named("invPred", pred && termToString$1(pred));
}
function predTriple(pred) {
    return new BasicLensM(({ quads, id }) => {
        const out = quads.filter((q) => q.subject.equals(id) && (!pred || q.predicate.equals(pred)));
        return out.map((q) => ({ quads, id: q }));
    }).named("predTriple");
}
function unique() {
    return new BasicLensM((qs) => {
        const literals = {};
        const named = {};
        const blank = {};
        for (const q of qs) {
            const ty = q.id.termType;
            if (ty === "Literal")
                literals[q.id.value] = q;
            if (ty === "NamedNode")
                named[q.id.value] = q;
            if (ty === "BlankNode")
                blank[q.id.value] = q;
        }
        const out = [];
        out.push(...Object.values(literals));
        out.push(...Object.values(named));
        out.push(...Object.values(blank));
        return out;
    }).named("unique");
}
function subjects() {
    return new BasicLensM((quads) => {
        return quads.map((x) => ({ id: x.subject, quads }));
    }).named("subjects");
}
function match(subject, predicate, object) {
    return new BasicLensM((quads) => {
        return quads
            .filter((x) => (!subject || x.subject.equals(subject)) &&
            (!predicate || x.predicate.equals(predicate)) &&
            (!object || x.object.equals(object)))
            .map((id) => ({ id, quads }));
    }).named("match", {
        subject: subject && termToString$1(subject),
        predicate: predicate && termToString$1(predicate),
        object: object && termToString$1(object),
    });
}
const subject = new BasicLens(({ id, quads }) => ({
    id: id.subject,
    quads,
})).named("subject");
const predicate = new BasicLens(({ id, quads }) => ({
    id: id.predicate,
    quads,
})).named("predicate");
const object = new BasicLens(({ id, quads }) => ({
    id: id.object,
    quads,
})).named("object");
function empty() {
    return new BasicLens((x) => x);
}

const RDFS = types.createTermNamespace("http://www.w3.org/2000/01/rdf-schema#", "subClassOf");
const SHACL = types.createTermNamespace("http://www.w3.org/ns/shacl#", "Shape", "NodeShape", "PropertyShape", "targetNode", "targetClass", "targetSubjectsOf", "targetObjectsOf", "property", "path", "class", "name", "description", "defaultValue", "alternativePath", "zeroOrMorePath", "oneOrMorePath", "zeroOrOnePath", "inversePath", "minCount", "maxCount", "datatype");
const RDFL = types.createUriAndTermNamespace("https://w3id.org/rdf-lens/ontology#", "CBD", "PathLens", "Context", "TypedExtract", "EnvVariable", "envKey", "envDefault", "datatype");

const { literal, quad } = new rdfDataFactory.DataFactory();
function termToString(term) {
    if (term.termType === "NamedNode") {
        return "<" + term.value + ">";
    }
    if (term.termType === "BlankNode") {
        return "_:" + term.value;
    }
    return JSON.stringify(term.value);
}
function fieldToLens(field) {
    const minCount = field.minCount || 0;
    const maxCount = field.maxCount || Number.MAX_SAFE_INTEGER;
    if (maxCount < 2) {
        return field.path.one(undefined).then(new BasicLens((x, ctx) => {
            if (x) {
                return field.extract.execute(x, ctx);
            }
            else {
                if (minCount > 0) {
                    throw new LensError("Field is not defined and required", ctx.lineage.slice());
                }
                else {
                    return x;
                }
            }
        }));
    }
    if (maxCount < 2)
        return field.path.one().then(field.extract);
    const thenListExtract = RdfList.and(empty()).map(([terms, { quads }]) => terms.map((id) => ({ id, quads })));
    const noListExtract = empty().map((x) => [x]);
    return field.path
        .thenFlat(thenListExtract.or(noListExtract).asMulti())
        .thenAll(field.extract)
        .map((x) => x.filter((x) => x !== undefined))
        .map((xs, ctx) => {
        if (xs.length < minCount) {
            throw new LensError("Mininum Count violation", [
                { name: "found:", opts: xs.length },
                ...ctx.lineage.slice(),
            ]);
        }
        if (xs.length > maxCount) {
            throw new LensError("Maximum Count violation", [
                { name: "found: " + xs.length, opts: [] },
                ...ctx.lineage.slice(),
            ]);
        }
        return xs;
    })
        .map((x) => {
        const out = x.filter((x) => x !== undefined);
        if (maxCount < 2) {
            return out[0];
        }
        else {
            return out;
        }
    });
}
function toLens(shape) {
    if (shape.fields.length === 0)
        return empty()
            .map(() => ({}))
            .named("first", shape.ty.value)
            .named("shape", {
            id: shape.id,
            type: termToString(shape.ty),
            description: shape.description,
        })
            .named("id", [], (cont) => termToString(cont.id));
    const fields = shape.fields.map((field) => {
        const base = fieldToLens(field);
        const asField = empty()
            .named("processing field", {
            name: field.name,
            minCount: field.minCount,
            maxCount: field.maxCount,
        })
            .then(base)
            .map((x) => {
            const out = {};
            out[field.name] = x;
            return out;
        });
        return asField;
    });
    return fields[0]
        .and(...fields.slice(1))
        .map((xs) => Object.assign({}, ...xs))
        .named("shape", {
        id: shape.id,
        type: termToString(shape.ty),
        description: shape.description,
    })
        .named("id", [], (cont) => termToString(cont.id));
}
const RDFListElement = pred(types.RDF.terms.first)
    .expectOne()
    .and(pred(types.RDF.terms.rest).expectOne());
const RdfList = new BasicLens((c, ctx) => {
    if (c.id.equals(types.RDF.terms.nil)) {
        return [];
    }
    const [first, rest] = RDFListElement.execute(c, ctx);
    const els = RdfList.execute(rest, ctx);
    els.unshift(first.id);
    return els;
});
const ShaclSequencePath = new BasicLens((c, ctx) => {
    const pathList = RdfList.execute(c, ctx);
    const paths = pathList.map((x) => ShaclPath.execute({ id: x, quads: c.quads }, ctx));
    if (paths.length === 0) {
        return new BasicLensM((c) => [c]);
    }
    let start = paths[0];
    for (let i = 1; i < pathList.length; i++) {
        start = start.thenFlat(paths[i]);
    }
    return start;
});
const ShaclAlternativepath = new BasicLens((c, ctx) => {
    const options = pred(SHACL.alternativePath)
        .one()
        .then(RdfList)
        .execute(c, ctx);
    const optionLenses = options.map((id) => ShaclPath.execute({ id, quads: c.quads }, ctx));
    return optionLenses[0].orAll(...optionLenses.slice(1));
});
const ShaclPredicatePath = extractLeaf(types.XSD.terms.custom("iri")).map(pred);
const ShaclInversePath = pred(SHACL.inversePath)
    .one()
    .then(new BasicLens((c, ctx) => {
    const pathList = RdfList.execute(c, ctx);
    if (pathList.length === 0) {
        return new BasicLensM((c) => [c]);
    }
    pathList.reverse();
    let start = invPred(pathList[0]);
    for (let i = 1; i < pathList.length; i++) {
        start = start.thenFlat(invPred(pathList[i]));
    }
    return start;
}).or(new BasicLens((c) => {
    return invPred(c.id);
})));
function MultiPath(predicate, min, max) {
    return pred(predicate)
        .one()
        .then(new BasicLens((c, ctx) => {
        return ShaclPath.execute(c, ctx);
    }))
        .map((x) => new BasicLensM((c, ctx) => {
        const out = [];
        let current = [c];
        let done = 0;
        if (min == 0) {
            out.push(c);
        }
        while (current.length > 0) {
            done += 1;
            const todo = current.slice();
            current = [];
            for (const c of todo) {
                try {
                    const news = x.execute(c, ctx);
                    current.push(...news);
                    if (done >= min && (!max || done <= max)) {
                        out.push(c);
                    }
                }
                catch (ex) {
                    console.log(ex);
                    if (done >= min && (!max || done <= max)) {
                        out.push(c);
                    }
                    break;
                }
            }
        }
        return out;
    }));
}
const ShaclPath = ShaclSequencePath.or(ShaclAlternativepath, ShaclInversePath, MultiPath(SHACL.zeroOrMorePath, 0), MultiPath(SHACL.zeroOrMorePath, 1), MultiPath(SHACL.zeroOrOnePath, 0, 1), ShaclPredicatePath);
function field(predicate, name, convert) {
    const conv = ((x) => x);
    return pred(predicate)
        .one()
        .map(({ id }) => {
        const out = {};
        out[name] = conv(id.value);
        return out;
    });
}
empty().map(({ id }) => id);
function optionalField(predicate, name, convert) {
    const conv = convert || ((x) => x);
    return pred(predicate)
        .one(undefined)
        .map((inp) => {
        const out = {};
        if (inp) {
            out[name] = conv(inp.id.value);
        }
        return out;
    });
}
function dataTypeToExtract(dataType, t) {
    if (dataType.equals(types.XSD.terms.integer))
        return +t.value;
    if (dataType.equals(types.XSD.terms.custom("float")))
        return +t.value;
    if (dataType.equals(types.XSD.terms.custom("double")))
        return +t.value;
    if (dataType.equals(types.XSD.terms.custom("decimal")))
        return +t.value;
    if (dataType.equals(types.XSD.terms.string))
        return t.value;
    if (dataType.equals(types.XSD.terms.dateTime))
        return new Date(t.value);
    if (dataType.equals(types.XSD.terms.custom("boolean")))
        return t.value === "true";
    if (dataType.equals(types.XSD.terms.custom("iri")))
        return new rdfDataFactory.NamedNode(t.value);
    if (dataType.equals(types.XSD.terms.custom("anyURI"))) {
        return new rdfDataFactory.NamedNode(t.value);
    }
    return t;
}
function envLens(dataType) {
    const checkType = pred(types.RDF.terms.type)
        .thenSome(new BasicLens(({ id }, ctx) => {
        if (!id.equals(RDFL.terms.EnvVariable)) {
            throw new LensError("Expected type " + RDFL.EnvVariable, ctx.lineage);
        }
        return { checked: true };
    }))
        .expectOne();
    const envName = pred(RDFL.terms.envKey)
        .one()
        .map(({ id }) => ({
        key: id.value,
    }));
    const defaultValue = pred(RDFL.terms.envDefault)
        .one(undefined)
        .map((found) => ({
        defaultValue: found?.id.value,
    }));
    const envDatatype = pred(RDFL.terms.datatype)
        .one(undefined)
        .map((found) => ({ dt: found?.id }));
    return checkType
        .and(envName, defaultValue, envDatatype)
        .map(([_, { key }, { defaultValue }, { dt }], ctx) => {
        const value = process.env[key] || defaultValue;
        const thisDt = dataType || dt || types.XSD.terms.custom("literal");
        if (value) {
            return dataTypeToExtract(thisDt, literal(value));
        }
        else {
            throw new LensError("ENV and default are not set", [
                { name: "Env Key", opts: key },
                ...ctx.lineage,
            ]);
        }
    });
}
function sliced() {
    return new BasicLens((x) => x.slice());
}
function remove_cbd(quads, subject) {
    const toRemoves = [subject];
    while (toRemoves.length > 0) {
        const toRemove = toRemoves.pop();
        quads = quads.filter((q) => {
            if (q.subject.equals(toRemove)) {
                if (q.object.termType === "BlankNode") {
                    toRemoves.push(q.object);
                }
                return false;
            }
            else {
                return true;
            }
        });
    }
    return quads;
}
function envReplace() {
    const shouldReplace = empty()
        .map((x) => x[0])
        .then(envLens().and(empty().map((x) => x.id)))
        .map(([value, id]) => ({
        value,
        id,
    }));
    const reduce = shouldReplace
        .and(empty().map((x) => x[1]))
        .map(([{ value, id }, quads]) => {
        return remove_cbd(quads.map((q) => {
            if (q.object.equals(id)) {
                return quad(q.subject, q.predicate, value, q.graph);
            }
            else {
                return q;
            }
        }), id);
    });
    const actualReplace = match(undefined, types.RDF.terms.type, RDFL.terms.EnvVariable)
        .thenAll(subject)
        .reduce(reduce, empty());
    return sliced().then(actualReplace);
}
function extractLeaf(datatype) {
    return envLens(datatype).or(empty().map((item) => dataTypeToExtract(datatype, item.id)));
}
function extractProperty(cache, _subClasses, apply) {
    const pathLens = pred(SHACL.path)
        .one()
        .then(ShaclPath)
        .map((path) => ({
        path,
    }));
    const nameLens = field(SHACL.custom("codeIdentifier"), "name").or(field(SHACL.name, "name"));
    const minCount = optionalField(SHACL.minCount, "minCount", (x) => +x);
    const maxCount = optionalField(SHACL.maxCount, "maxCount", (x) => +x);
    const dataTypeLens = pred(SHACL.datatype)
        .one()
        .map(({ id }) => ({
        extract: extractLeaf(id),
    }));
    const clazzLens = field(SHACL.class, "clazz").map(({ clazz: expected_class }) => {
        return {
            extract: new BasicLens(({ id, quads }, ctx) => {
                const lens = cache[expected_class];
                if (!lens) {
                    throw new LensError("Tried extracting class, but no shape was defined", [
                        {
                            name: "Found type: " + expected_class,
                            opts: Object.keys(cache),
                        },
                        ...ctx.lineage.slice(),
                    ]);
                }
                if (apply[expected_class]) {
                    return lens
                        .map(apply[expected_class])
                        .execute({ id, quads }, ctx);
                }
                else {
                    return lens.execute({ id, quads }, ctx);
                }
            }).named("extracting class", expected_class),
        };
    });
    return pathLens
        .and(nameLens, minCount, maxCount, clazzLens.or(dataTypeLens))
        .map((xs) => Object.assign({}, ...xs));
}
const CBDLens = new BasicLensM(({ id, quads }, cont) => {
    cont.lineage.push({ name: "CBD", opts: ["from: " + id.value] });
    const done = new Set();
    const todo = [id];
    const out = [];
    let item = todo.pop();
    while (item) {
        const found = quads.filter((x) => x.subject.equals(item));
        out.push(...found);
        for (const option of found) {
            const object = option.object;
            if (object.termType !== "BlankNode") {
                continue;
            }
            if (done.has(object.value))
                continue;
            done.add(object.value);
            todo.push(object);
        }
        item = todo.pop();
    }
    return out;
});
function getCacheState(le, ctx, st) {
    const out = ctx.stateMap.get(le);
    if (out !== undefined)
        return out;
    const o = st();
    ctx.stateMap.set(le, o);
    return o;
}
const Cached = function (lens, cachedLenses) {
    const lenses = cachedLenses["lenses"] ?? (cachedLenses.lenses = []);
    const found = lenses.find((x) => x.from === lens);
    if (found) {
        return found.lens;
    }
    const newLens = new BasicLens(({ id, quads }, ctx) => {
        const state = getCacheState(newLens, ctx, () => ({
            namedNodes: {},
            blankNodes: {},
        }));
        let stateDict = {};
        if (id.termType == "NamedNode") {
            stateDict = state.namedNodes = state.namedNodes ?? {};
        }
        if (id.termType == "BlankNode") {
            stateDict = state.blankNodes = state.blankNodes ?? {};
        }
        if (!(id.value in stateDict)) {
            stateDict[id.value] = [];
        }
        const res = stateDict[id.value].find((x) => x.lens == lens);
        if (res) {
            return res.result;
        }
        const thisThing = { lens: lens, result: {} };
        stateDict[id.value].push(thisThing);
        const executedLens = lens.execute({ quads, id }, ctx);
        Object.assign(thisThing.result, executedLens);
        return thisThing.result;
    });
    lenses.push({ lens: newLens, from: lens });
    return newLens;
};
const TypedExtract = function (cache, apply, subClasses) {
    const lens = new BasicLens(({ id, quads }, ctx) => {
        const ty = quads.find((q) => q.subject.equals(id) && q.predicate.equals(types.RDF.terms.type))?.object.value;
        ctx.lineage.push({ name: "Found type", opts: ty });
        ctx.lineage.push({ name: "TypedExtract", opts: undefined });
        if (!ty) {
            throw new LensError("Expected a type, found none", ctx.lineage.slice());
        }
        const lenses = [];
        let current = ty;
        while (current) {
            const thisLens = cache[current];
            if (thisLens) {
                const state = getCacheState(lens, ctx, () => ({
                    lenses: [],
                }));
                lenses.push(Cached(thisLens, state));
            }
            current = subClasses[current];
        }
        if (lenses.length === 0) {
            throw new LensError("Expected a lens for type, found none", ctx.lineage.slice());
        }
        const finalLens = lenses.length == 1
            ? lenses[0]
            : lenses[0]
                .and(...lenses.slice(1))
                .map((xs) => Object.assign({}, ...xs));
        if (apply[ty]) {
            return finalLens.map(apply[ty]).execute({ id, quads }, ctx);
        }
        else {
            return finalLens.execute({ id, quads }, ctx);
        }
    });
    return lens;
};
function extractShape(cache, subclasses, apply) {
    const checkTy = pred(types.RDF.terms.type)
        .one()
        .map(({ id }, ctx) => {
        if (id.equals(SHACL.NodeShape))
            return {};
        throw new LensError("Expected type sh:NodeShape", [
            { name: "found type", opts: termToString(id) },
            ...ctx.lineage,
        ]);
    });
    const idLens = empty().map(({ id }) => ({ id: id.value }));
    const clazzs = pred(SHACL.targetClass);
    const multiple = clazzs.thenAll(empty().map(({ id }) => ({ ty: id })));
    const descriptionClassLens = optionalField(SHACL.description, "description");
    const fields = pred(SHACL.property)
        .thenAll(extractProperty(cache, subclasses, apply))
        .map((fields) => ({ fields }));
    return multiple
        .and(checkTy, idLens, descriptionClassLens, fields)
        .map(([multiple, ...others]) => multiple.map((xs) => Object.assign({}, xs, ...others)));
}
function extractShapes(quads, apply = {}, customClasses = {}) {
    const cache = Object.assign({}, customClasses);
    cache[RDFL.PathLens] = ShaclPath;
    cache[RDFL.CBD] = CBDLens;
    cache[RDFL.Context] = new BasicLens(({ quads }) => {
        return quads;
    });
    const subClasses = {};
    quads
        .filter((x) => x.predicate.equals(RDFS.subClassOf))
        .forEach((x) => (subClasses[x.subject.value] = x.object.value));
    const shapes = subjects()
        .then(unique())
        .asMulti()
        .thenSome(extractShape(cache, subClasses, apply))
        .execute(quads, createContext())
        .flat();
    cache[RDFL.TypedExtract] = TypedExtract(cache, apply, subClasses);
    for (const shape of shapes) {
        const lens = toLens(shape);
        const target = cache[shape.ty.value];
        if (target) {
            cache[shape.ty.value] = target.or(lens);
        }
        else {
            cache[shape.ty.value] = lens;
        }
    }
    return { lenses: cache, shapes, subClasses };
}

exports.BasicLens = BasicLens;
exports.BasicLensM = BasicLensM;
exports.CBDLens = CBDLens;
exports.Cached = Cached;
exports.LensError = LensError;
exports.MultiPath = MultiPath;
exports.RdfList = RdfList;
exports.ShaclAlternativepath = ShaclAlternativepath;
exports.ShaclInversePath = ShaclInversePath;
exports.ShaclPath = ShaclPath;
exports.ShaclPredicatePath = ShaclPredicatePath;
exports.ShaclSequencePath = ShaclSequencePath;
exports.TypedExtract = TypedExtract;
exports.createContext = createContext;
exports.empty = empty;
exports.envReplace = envReplace;
exports.extractShape = extractShape;
exports.extractShapes = extractShapes;
exports.invPred = invPred;
exports.match = match;
exports.object = object;
exports.pred = pred;
exports.predTriple = predTriple;
exports.predicate = predicate;
exports.sliced = sliced;
exports.subject = subject;
exports.subjects = subjects;
exports.toLens = toLens;
exports.unique = unique;
