// rdf-lens/dist/lens.js
function termToString(term) {
    if (term.termType === "NamedNode") {
        return "<" + term.value + ">";
    }
    if (term.termType === "BlankNode") {
        return "_:" + term.value;
    }
    return JSON.stringify(term.value);
}
export class LensError extends Error {
    lineage;
    constructor(message, lineage) {
        super(message);
        this.message = message;
        this.lineage = lineage;
    }
}
export function createContext() {
    const ctx = {
        stateMap: new Map(),
        lineage: [],
    };
    const clone = () => ({
        clone,
        stateMap: ctx.stateMap,
        lineage: ctx.lineage.slice(),
    });
    return Object.assign(ctx, { clone });
}
export class BasicLens {
    _exec;
    index;
    constructor(execute) {
        this._exec = execute;
    }
    named(name, opts, cb) {
        return new BasicLens((c, ctx) => {
            let extras = asList(opts) || [];
            if (cb) {
                extras = [...extras, ...asList(cb(c))];
            }
            ctx.lineage.push({ name, opts: deconstructList(extras) });
            return this.execute(c, ctx);
        });
    }
    asMulti() {
        return (new BasicLensM((c, ctx) => {
            const out = this.execute(c, ctx);
            return out;
        }));
    }
    and(...and) {
        return (new BasicLens((c, ctx) => {
            const a = this.execute(c, ctx);
            const rest = and.map((x) => x.execute(c, ctx));
            return [a, ...rest];
        }));
    }
    orM(...others) {
        return new BasicLensM((c, ctx) => {
            const all = [this, ...others];
            return all.flatMap((x) => {
                try {
                    return [x.execute(c, ctx.clone())];
                }
                catch (ex) {
                    return [];
                }
            });
        });
    }
    or(...others) {
        return new BasicLens((c, ctx) => {
            const errors = [];
            try {
                return this.execute(c, ctx);
            }
            catch (ex) {
                errors.push(ex);
                for (let i = 0; i < others.length; i++) {
                    try {
                        return others[i].execute(c, ctx.clone());
                    }
                    catch (ex) {
                        errors.push(ex);
                    }
                }
            }
            throw errors;
        });
    }
    map(fn) {
        return new BasicLens((c, ctx) => {
            const a = this.execute(c, ctx);
            return fn(a, ctx);
        });
    }
    then(next) {
        return new BasicLens((c, ctx) => {
            const a = this.execute(c, ctx);
            return next.execute(a, ctx);
        });
    }
    execute(container, ctx = createContext()) {
        return this._exec(container, ctx);
    }
}
function deconstructList(x) {
    if (x.length == 1) {
        return x[0];
    }
    return x;
}
function asList(x) {
    if (Array.isArray(x))
        return x;
    return [x];
}
export class BasicLensM extends BasicLens {
    named(name, opts, cb) {
        return new BasicLensM((c, ctx) => {
            let extras = asList(opts) || [];
            if (cb) {
                extras = [...extras, ...asList(cb(c))];
            }
            ctx.lineage.push({ name, opts: deconstructList(extras) });
            return this.execute(c, ctx);
        });
    }
    one(def) {
        return new BasicLens((c, ctx) => {
            const qs = this.execute(c, ctx);
            return qs[0] || def;
        });
    }
    expectOne() {
        return new BasicLens((c, ctx) => {
            const qs = this.execute(c, ctx);
            if (qs.length < 1)
                throw new LensError("Expected one, found none", ctx.lineage.slice());
            return qs[0];
        });
    }
    thenAll(next) {
        return new BasicLensM((c, ctx) => {
            const qs = this.execute(c, ctx.clone());
            return qs.map((x) => next.execute(x, ctx.clone()));
        });
    }
    thenSome(next) {
        return new BasicLensM((c, ctx) => {
            const qs = this.execute(c, ctx.clone());
            return qs.flatMap((x) => {
                try {
                    const o = next.execute(x, ctx.clone());
                    return [o];
                }
                catch (ex) {
                    return [];
                }
            });
        });
    }
    thenFlat(next) {
        return new BasicLensM((c, ctx) => {
            const qs = this.execute(c, ctx.clone());
            return qs.flatMap((x) => next.execute(x, ctx.clone()));
        });
    }
    mapAll(fn) {
        return new BasicLensM((c, ctx) => {
            const qs = this.execute(c, ctx);
            return qs.map((x) => fn(x, ctx));
        });
    }
    orAll(...others) {
        return new BasicLensM((c, ctx) => {
            const out = [];
            try {
                out.push(...this.execute(c, ctx.clone()));
            }
            catch (ex) {
            }
            for (let i = 0; i < others.length; i++) {
                try {
                    out.push(...others[i].execute(c, ctx.clone()));
                }
                catch (ex) {
                }
            }
            return out;
        });
    }
    filter(fn) {
        return new BasicLensM((c, ctx) => {
            return this.execute(c, ctx).filter(fn);
        });
    }
    reduce(lens, start) {
        return new BasicLens((c, ctx) => {
            const st = this.and(start).map(([ts, f]) => {
                return ts.reduce((acc, v) => lens.execute([v, acc], ctx), f);
            });
            return st.execute(c, ctx);
        });
    }
}
export function pred(pred) {
    return new BasicLensM(({ quads, id }) => {
        const out = quads.filter((q) => q.subject.equals(id) && (!pred || q.predicate.equals(pred)));
        return out.map((q) => ({ quads, id: q.object }));
    }).named("pred", pred && termToString(pred));
}
export function invPred(pred) {
    return new BasicLensM(({ quads, id }) => {
        const out = quads.filter((q) => q.object.equals(id) && (!pred || q.predicate.equals(pred)));
        return out.map((q) => ({ quads, id: q.subject }));
    }).named("invPred", pred && termToString(pred));
}
export function predTriple(pred) {
    return new BasicLensM(({ quads, id }) => {
        const out = quads.filter((q) => q.subject.equals(id) && (!pred || q.predicate.equals(pred)));
        return out.map((q) => ({ quads, id: q }));
    }).named("predTriple");
}
export function unique() {
    return new BasicLensM((qs) => {
        const literals = {};
        const named = {};
        const blank = {};
        for (const q of qs) {
            const ty = q.id.termType;
            if (ty === "Literal")
                literals[q.id.value] = q;
            if (ty === "NamedNode")
                named[q.id.value] = q;
            if (ty === "BlankNode")
                blank[q.id.value] = q;
        }
        const out = [];
        out.push(...Object.values(literals));
        out.push(...Object.values(named));
        out.push(...Object.values(blank));
        return out;
    }).named("unique");
}
export function subjects() {
    return new BasicLensM((quads) => {
        return quads.map((x) => ({ id: x.subject, quads }));
    }).named("subjects");
}
export function match(subject, predicate, object) {
    return new BasicLensM((quads) => {
        return quads
            .filter((x) => (!subject || x.subject.equals(subject)) &&
            (!predicate || x.predicate.equals(predicate)) &&
            (!object || x.object.equals(object)))
            .map((id) => ({ id, quads }));
    }).named("match", {
        subject: subject && termToString(subject),
        predicate: predicate && termToString(predicate),
        object: object && termToString(object),
    });
}
export const subject = new BasicLens(({ id, quads }) => ({
    id: id.subject,
    quads,
})).named("subject");
export const predicate = new BasicLens(({ id, quads }) => ({
    id: id.predicate,
    quads,
})).named("predicate");
export const object = new BasicLens(({ id, quads }) => ({
    id: id.object,
    quads,
})).named("object");
export function empty() {
    return new BasicLens((x) => x);
}
