// @rdfc/sds-processors-ts/lib/bucketizers/dumpBucketizer.js
import { getLoggerFor } from "../utils/logUtil.js";
import { findAndValidateRecordTimestamp, } from "../utils/index.js";
export default class DumpBucketizer {
    logger = getLoggerFor(this);
    path;
    pathQuads;
    lastMemberTimestamp = 0;
    constructor(config, save) {
        if (config.path && config.pathQuads) {
            this.path = config.path.mapAll((x) => ({
                value: x.id.value,
                literal: x.id,
            }));
            this.pathQuads = config.pathQuads;
        }
        if (save) {
            const parsed = JSON.parse(save);
            this.lastMemberTimestamp = parsed.lastMemberTimestamp;
        }
    }
    bucketize(record, getBucket) {
        const recordTimestamp = findAndValidateRecordTimestamp(record, this.path, this.lastMemberTimestamp, this.logger);
        if (recordTimestamp === null) {
            return [];
        }
        if (recordTimestamp) {
            this.lastMemberTimestamp = recordTimestamp.getTime();
        }
        const bucket = getBucket("", true);
        return [bucket];
    }
    save() {
        return JSON.stringify({
            lastMemberTimestamp: this.lastMemberTimestamp,
        });
    }
}
