// @rdfc/sds-processors-ts/lib/bucketizers/hourBucketizer.js
import { getLoggerFor } from "../utils/logUtil.js";
import { TREE, XSD } from "@treecg/types";
import { DataFactory } from "n3";
const { literal, namedNode } = DataFactory;
export default class HourBucketizer {
    logger = getLoggerFor(this);
    path;
    pathQuads;
    unorderedRelations;
    hour;
    root = true;
    constructor(config, save) {
        this.path = config.path.mapAll((x) => ({
            value: x.id.value,
            literal: x.id,
        }));
        this.pathQuads = config.pathQuads;
        this.unorderedRelations = config.unorderedRelations ?? false;
        if (save) {
            const parsed = JSON.parse(save);
            this.hour = new Date(parsed.hour);
            this.root = parsed.root;
        }
    }
    bucketize(record, getBucket, addRelation) {
        const values = this.path
            .execute(record.data)
            .filter((x, i, arr) => arr.findIndex((y) => x.value === y.value) == i);
        const out = [];
        for (const value of values) {
            if (value.literal) {
                const timestamp = value.literal.value;
                const recordDate = new Date(timestamp);
                if (!this.hour) {
                    this.root = true;
                    this.hour = new Date(recordDate.getFullYear(), recordDate.getMonth(), recordDate.getDate(), recordDate.getHours());
                    out.push(getBucket(this.hour.toISOString(), this.root));
                    this.logger.debug(`Created root hour bucket ${this.hour.toISOString()}`);
                }
                else if (recordDate.getHours() !== this.hour.getHours()) {
                    const newHour = new Date(recordDate.getFullYear(), recordDate.getMonth(), recordDate.getDate(), recordDate.getHours());
                    const newBucket = getBucket(newHour.toISOString(), false);
                    const oldBucket = getBucket(this.hour.toISOString(), this.root);
                    this.root = false;
                    if (this.unorderedRelations) {
                        addRelation(oldBucket, newBucket, TREE.terms.Relation);
                        addRelation(newBucket, oldBucket, TREE.terms.Relation);
                    }
                    else {
                        addRelation(oldBucket, newBucket, TREE.terms.GreaterThanOrEqualToRelation, literal(this.hour.toISOString(), namedNode(XSD.dateTime)), this.pathQuads);
                        addRelation(newBucket, oldBucket, TREE.terms.LessThanRelation, literal(this.hour.toISOString(), namedNode(XSD.dateTime)), this.pathQuads);
                    }
                    oldBucket.immutable = true;
                    out.push(newBucket);
                    this.hour = newHour;
                    this.logger.debug(`Created new hour bucket ${this.hour.toISOString()}`);
                }
                else {
                    out.push(getBucket(this.hour.toISOString(), this.root));
                }
            }
            else {
                this.logger.warn(`Received records without timestamp values. Ignoring record '${record.data.id.value}'.`);
            }
        }
        return out;
    }
    save() {
        return JSON.stringify({
            hour: this.hour,
            root: this.root,
        });
    }
}
