// @rdfc/sds-processors-ts/lib/bucketizers/index.js
import { Bucket, getOrDefaultMap, } from "../utils/index.js";
import { TREE } from "@treecg/types";
import { DataFactory } from "rdf-data-factory";
import PagedBucketizer from "./pagedBucketizer.js";
import SubjectBucketizer from "./subjectBucketizer.js";
import TimebasedBucketizer from "./timebasedBucketizer.js";
import TimeBucketBucketizer from "./timeBucketTree.js";
import HourBucketizer from "./hourBucketizer.js";
import ReversedPagedBucketizer from "./reversedPagedBucketizer.js";
import DumpBucketizer from "./dumpBucketizer.js";
const df = new DataFactory();
export const SHAPES_TEXT = "@prefix tree: <https://w3id.org/tree#>.\n@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.\n@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#>.\n@prefix sds: <https://w3id.org/sds#>.\n@prefix sh: <http://www.w3.org/ns/shacl#>.\n@prefix xsd: <http://www.w3.org/2001/XMLSchema#>.\n@prefix rdfl: <https://w3id.org/rdf-lens/ontology#>.\n\n[ ] a sh:NodeShape;\n  sh:targetClass tree:FragmentationStrategy;\n  sh:property [\n    sh:name \"type\";\n    sh:path rdf:type;\n    sh:datatype xsd:iri;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"config\";\n    sh:path ( );\n    sh:class rdfl:TypedExtract;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ].\n\n[ ] a sh:NodeShape;\n  sh:targetClass <RdfThing>;\n  sh:property [\n    sh:name \"id\";\n    sh:path ( );\n    sh:maxCount 1;\n    sh:minCount 1;\n    sh:datatype xsd:any;\n  ], [\n    sh:name \"quads\";\n    sh:path ( );\n    sh:maxCount 1;\n    sh:minCount 1;\n    sh:class rdfl:CBD;\n  ].\n\n[ ] a sh:NodeShape;\n  sh:targetClass tree:SubjectFragmentation;\n  sh:property [\n    sh:name \"path\";\n    sh:path tree:fragmentationPath;\n    sh:class rdfl:PathLens;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"pathQuads\";\n    sh:path tree:fragmentationPath;\n    sh:class <RdfThing>;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"namePath\";\n    sh:path tree:fragmentationPathName;\n    sh:class rdfl:PathLens;\n    sh:maxCount 1;\n  ], [\n    sh:name \"defaultName\";\n    sh:path tree:fragmentationDefaultName;\n    sh:datatype xsd:string;\n    sh:maxCount 1;\n  ].\n\n[ ] a sh:NodeShape;\n  sh:targetClass tree:PageFragmentation;\n  sh:property [\n    sh:name \"pageSize\";\n    sh:path tree:pageSize;\n    sh:datatype xsd:integer;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"path\";\n    sh:path tree:timestampPath;\n    sh:class rdfl:PathLens;\n    sh:maxCount 1;\n    sh:minCount 0;\n  ], [\n    sh:name \"pathQuads\";\n    sh:path tree:timestampPath;\n    sh:class <RdfThing>;\n    sh:maxCount 1;\n    sh:minCount 0;\n  ].\n\n[ ] a sh:NodeShape;\n    sh:targetClass tree:ReversedPageFragmentation;\n    sh:property [\n        sh:name \"pageSize\";\n        sh:path tree:pageSize;\n        sh:datatype xsd:integer;\n        sh:maxCount 1;\n        sh:minCount 1;\n      ], [\n        sh:name \"path\";\n        sh:path tree:timestampPath;\n        sh:class rdfl:PathLens;\n        sh:maxCount 1;\n        sh:minCount 0;\n      ], [\n        sh:name \"pathQuads\";\n        sh:path tree:timestampPath;\n        sh:class <RdfThing>;\n        sh:maxCount 1;\n        sh:minCount 0;\n      ].\n\n[ ] a sh:NodeShape;\n  sh:targetClass tree:TimebasedFragmentation;\n  sh:property [\n    sh:name \"path\";\n    sh:path tree:timestampPath;\n    sh:class rdfl:PathLens;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"pathQuads\";\n    sh:path tree:timestampPath;\n    sh:class <RdfThing>;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"maxSize\";\n    sh:path tree:maxSize;\n    sh:datatype xsd:integer;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"k\";\n    sh:path tree:k;\n    sh:datatype xsd:integer;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"minBucketSpan\";\n    sh:path tree:minBucketSpan;\n    sh:datatype xsd:integer;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ].\n\n[ ] a sh:NodeShape;\n  sh:property [\n    sh:name \"ranges\";\n    sh:path tree:range;\n    sh:minCount 1;\n    sh:datatype xsd:string;\n    sh:in ( \"year\" \"month\" \"day-of-month\" \"hour\" \"minute\" \"second\" \"millisecond\" );\n  ], [\n    sh:name \"amount\";\n    sh:path tree:maxSize;\n    sh:maxCount 1;\n    sh:minCount 1;\n    sh:datatype xsd:integer;\n  ].\n\n[ ] a sh:NodeShape;\n  sh:targetClass tree:TimeBucketLevel;\n  sh:property [\n    sh:name \"ranges\";\n    sh:path tree:range;\n    sh:minCount 1;\n    sh:datatype xsd:string;\n    sh:in ( \"year\" \"month\" \"day-of-month\" \"hour\" \"minute\" \"second\" \"millisecond\" );\n  ], [\n    sh:name \"amount\";\n    sh:path tree:maxSize;\n    sh:maxCount 1;\n    sh:minCount 1;\n    sh:datatype xsd:integer;\n  ].\n\n[ ] a sh:NodeShape;\n  sh:targetClass tree:TimeBucketFragmentation;\n  sh:property [\n    sh:name \"path\";\n    sh:path tree:timestampPath;\n    sh:class rdfl:PathLens;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"pathQuads\";\n    sh:path tree:timestampPath;\n    sh:class <RdfThing>;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"levels\";\n    sh:path tree:level;\n    sh:class tree:TimeBucketLevel;\n    sh:minCount 1;\n  ], [\n    sh:name \"timeBufferMs\";\n    sh:path tree:buffer;\n    sh:datatype xsd:integer;\n    sh:maxCount 1;\n  ].\n\n[ ] a sh:NodeShape;\n  sh:targetClass tree:HourFragmentation;\n  sh:property [\n    sh:name \"path\";\n    sh:path tree:timestampPath;\n    sh:class rdfl:PathLens;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"pathQuads\";\n    sh:path tree:timestampPath;\n    sh:class <RdfThing>;\n    sh:maxCount 1;\n    sh:minCount 1;\n  ], [\n    sh:name \"unorderedRelations\";\n    sh:path tree:unorderedRelations;\n    sh:datatype xsd:boolean;\n    sh:maxCount 1;\n    sh:minCount 0;\n  ].\n\n[ ] a sh:NodeShape;\n  sh:targetClass tree:DumpFragmentation;\n  sh:property [\n    sh:name \"path\";\n    sh:path tree:timestampPath;\n    sh:class rdfl:PathLens;\n    sh:maxCount 1;\n    sh:minCount 0;\n  ], [\n    sh:name \"pathQuads\";\n    sh:path tree:timestampPath;\n    sh:class <RdfThing>;\n    sh:maxCount 1;\n    sh:minCount 0;\n  ].\n";
function createBucketizer(config, save) {
    switch (config.type.value) {
        case TREE.custom("SubjectFragmentation"):
            return new SubjectBucketizer(config.config, save);
        case TREE.custom("PageFragmentation"):
            return new PagedBucketizer(config.config, save);
        case TREE.custom("ReversedPageFragmentation"):
            return new ReversedPagedBucketizer(config.config, save);
        case TREE.custom("TimebasedFragmentation"):
            return new TimebasedBucketizer(config.config, save);
        case TREE.custom("TimeBucketFragmentation"):
            return new TimeBucketBucketizer(config.config, save);
        case TREE.custom("HourFragmentation"):
            return new HourBucketizer(config.config, save);
        case TREE.custom("DumpFragmentation"):
            return new DumpBucketizer(config.config, save);
    }
    throw "Unknown bucketizer " + config.type.value;
}
function combineIds(id1, id2) {
    const id1Slash = id1.endsWith("/");
    const id2Slash = id1.startsWith("/");
    if (id1Slash && id2Slash)
        return id1 + id2.slice(1);
    if (id1 === "" || id1Slash || id2Slash)
        return id1 + id2;
    return id1 + "/" + id2;
}
export class BucketizerOrchestrator {
    configs;
    saves = {};
    constructor(configs, save) {
        this.configs = configs;
        if (save) {
            this.saves = JSON.parse(save);
        }
    }
    bucketize(record, buckets, requestedBuckets, newMembers, newRelations, removeRelations, prefix) {
        let queue = [prefix];
        const addRelation = (origin, target, type, value, path) => {
            const relation = {
                type,
                value,
                path,
                target: target.id,
            };
            const newRel = {
                origin,
                relation,
            };
            newRelations.push(newRel);
            origin.links.push(relation);
            target.parent = origin;
        };
        const removeRelation = (origin, target, type, value, path) => {
            const relation = {
                type,
                value,
                path,
                target: target.id,
            };
            const removeRel = {
                origin,
                relation,
            };
            removeRelations.push(removeRel);
            origin.links = origin.links.filter((x) => !(x.type.equals(type) &&
                x.target.equals(target.id) &&
                (!value || x.value?.equals(value)) &&
                (!path || x.path?.id.equals(path.id))));
            target.parent = undefined;
        };
        for (let i = 0; i < this.configs.length; i++) {
            const todo = queue.slice();
            queue = [];
            for (const prefix of todo) {
                const bucketizer = this.getBucketizer(i, prefix);
                const getBucket = (value, root, keyIsId = false) => {
                    const encodedValue = value
                        .split("/")
                        .map((x) => encodeURIComponent(decodeURIComponent(x)))
                        .join("/");
                    const key = value.endsWith("/")
                        ? encodedValue
                        : encodedValue + "/";
                    const next = combineIds(prefix, key);
                    const id = keyIsId ? key : root ? prefix : next;
                    if (!buckets[id]) {
                        buckets[id] = new Bucket(df.namedNode(id), [], false);
                        buckets[id].addMember = (memberId) => {
                            getOrDefaultMap(newMembers, id, new Set()).add(memberId);
                        };
                    }
                    requestedBuckets.add(id);
                    return buckets[id];
                };
                const foundBucket = bucketizer.bucketize(record, getBucket, addRelation, removeRelation);
                for (const bucket of foundBucket) {
                    queue.push(bucket.id.value);
                }
            }
        }
        if (buckets[prefix]) {
            buckets[prefix].root = true;
        }
        return queue;
    }
    save() {
        for (const key of Object.keys(this.saves)) {
            if (this.saves[key].bucketizer) {
                this.saves[key].save = this.saves[key].bucketizer?.save();
                delete this.saves[key].bucketizer;
            }
        }
        return JSON.stringify(this.saves);
    }
    getBucketizer(index, key) {
        if (!this.saves[key]) {
            this.saves[key] = {};
        }
        const save = this.saves[key];
        if (!save.bucketizer) {
            save.bucketizer = createBucketizer(this.configs[index], save.save);
        }
        return save.bucketizer;
    }
}
