// @rdfc/sds-processors-ts/lib/bucketizers/pagedBucketizer.js
import { findAndValidateRecordTimestamp, } from "../utils/index.js";
import { TREE, XSD } from "@treecg/types";
import { getLoggerFor } from "../utils/logUtil.js";
import { DataFactory } from "n3";
const { literal, namedNode } = DataFactory;
export default class PagedBucketizer {
    logger = getLoggerFor(this);
    pageSize;
    path;
    pathQuads;
    count = 0;
    lastMemberTimestamp = 0;
    constructor(config, save) {
        this.pageSize = config.pageSize;
        if (config.path && config.pathQuads) {
            this.path = config.path.mapAll((x) => ({
                value: x.id.value,
                literal: x.id,
            }));
            this.pathQuads = config.pathQuads;
        }
        if (save) {
            const parsed = JSON.parse(save);
            this.count = parsed.count;
            this.lastMemberTimestamp = parsed.lastMemberTimestamp;
        }
    }
    bucketize(record, getBucket, addRelation, removeRelation) {
        const index = Math.floor(this.count / this.pageSize);
        this.count += 1;
        const currentBucket = getBucket("page-" + index, index == 0);
        const recordTimestamp = findAndValidateRecordTimestamp(record, this.path, this.lastMemberTimestamp, this.logger);
        if (recordTimestamp === null) {
            return [];
        }
        if (recordTimestamp) {
            this.lastMemberTimestamp = recordTimestamp.getTime();
        }
        if (this.count % this.pageSize == 1 && this.count > 1) {
            const oldBucket = getBucket("page-" + (index - 1), index - 1 == 0);
            oldBucket.immutable = true;
            if (recordTimestamp) {
                addRelation(oldBucket, currentBucket, TREE.terms.GreaterThanOrEqualToRelation, literal(recordTimestamp.toISOString(), namedNode(XSD.dateTime)), this.pathQuads);
            }
            else {
                addRelation(oldBucket, currentBucket, TREE.terms.Relation);
            }
            this.logger.info(`Created new page bucket ${index}`);
        }
        return [currentBucket];
    }
    save() {
        return JSON.stringify({
            count: this.count,
            lastMemberTimestamp: this.lastMemberTimestamp,
        });
    }
}
