// @rdfc/sds-processors-ts/lib/bucketizers/subjectBucketizer.js
import { TREE } from "@treecg/types";
import { getLoggerFor } from "../utils/logUtil.js";
export default class SubjectBucketizer {
    logger = getLoggerFor(this);
    path;
    pathQuads;
    namePath;
    defaultName;
    seen = new Set();
    constructor(config, save) {
        this.path = config.path.mapAll((x) => ({
            value: x.id.value,
            literal: x.id,
        }));
        this.pathQuads = config.pathQuads;
        this.namePath = config.namePath;
        this.defaultName = config.defaultName;
        if (save) {
            this.seen = new Set(JSON.parse(save));
        }
    }
    bucketize(record, getBucket, addRelation) {
        const values = this.path
            .execute(record.data)
            .filter((x, i, arr) => arr.findIndex((y) => x.value === y.value) == i);
        const out = [];
        const root = getBucket("", true);
        if (values.length === 0 && this.defaultName) {
            values.push({ value: this.defaultName });
        }
        if (values.length === 0) {
            this.logger.error("Didn't find bucket, and default name is not set, sadness :(");
        }
        for (const value of values) {
            const name = encodeURIComponent(value.literal
                ? this.namePath?.execute({
                    id: value.literal,
                    quads: record.data.quads,
                })[0]?.id.value || value.value
                : value.value);
            const bucket = getBucket(name);
            if (!this.seen.has(bucket.id.value)) {
                this.seen.add(bucket.id.value);
                addRelation(root, bucket, TREE.terms.EqualToRelation, value.literal, value.literal ? this.pathQuads : undefined);
            }
            out.push(bucket);
        }
        return out;
    }
    save() {
        return JSON.stringify([...this.seen.values()]);
    }
}
