// @rdfc/sds-processors-ts/lib/bucketizers/timebasedBucketizer.js
import { TREE, XSD } from "@treecg/types";
import { DataFactory } from "n3";
import { getLoggerFor } from "../utils/logUtil.js";
const { literal, namedNode } = DataFactory;
export default class TimebasedBucketizer {
    logger = getLoggerFor(this);
    path;
    pathQuads;
    maxSize = 100;
    k = 4;
    minBucketSpan = 300000;
    mutableLeafBucketKeys = [];
    members = [];
    constructor(config, save) {
        this.path = config.path.mapAll((x) => ({
            value: x.id.value,
            literal: x.id,
        }));
        this.pathQuads = config.pathQuads;
        this.maxSize = config.maxSize;
        this.k = config.k;
        this.minBucketSpan = config.minBucketSpan * 1000;
        if (save) {
            const parsed = JSON.parse(save);
            this.mutableLeafBucketKeys = parsed.mutableLeafBucketKeys;
            this.members = parsed.members;
        }
    }
    bucketize(record, getBucket, addRelation) {
        const values = this.path
            .execute(record.data)
            .filter((x, i, arr) => arr.findIndex((y) => x.value === y.value) == i);
        const out = [];
        for (const value of values) {
            if (value.literal) {
                const timestamp = value.literal.value;
                let candidateBucket = undefined;
                let bucketKey = this.mutableLeafBucketKeys[0];
                while (!candidateBucket) {
                    if (!bucketKey) {
                        const root = getBucket("", true);
                        const newBucketTimestamp = new Date(timestamp);
                        newBucketTimestamp.setUTCMonth(0);
                        newBucketTimestamp.setUTCDate(1);
                        newBucketTimestamp.setUTCHours(0, 0, 0, 0);
                        const yearTimespan = new Date(newBucketTimestamp.getUTCFullYear(), 1, 29).getDate() === 29
                            ? 31622400000
                            : 31536000000;
                        this.logger.info(`Creating new year bucket with timestamp ${newBucketTimestamp.toISOString()} and timespan ${yearTimespan}.`);
                        const yearBucket = getBucket(`${newBucketTimestamp.toISOString()}_${yearTimespan}_0`, false);
                        this.addTimestampRelations(root, yearBucket, newBucketTimestamp, yearTimespan, addRelation);
                        this.mutableLeafBucketKeys.push(yearBucket.id.value);
                        this.members = [];
                        bucketKey = yearBucket.id.value;
                    }
                    const bucket = getBucket(bucketKey, false, true);
                    const bucketNames = bucket.id.value.split("/");
                    const bucketProperties = decodeURIComponent(bucketNames[bucketNames.length - 2]).split("_");
                    const bucketTimestamp = new Date(bucketProperties[0]);
                    const bucketSpan = parseInt(bucketProperties[1]);
                    const recordTimestamp = new Date(timestamp);
                    if (recordTimestamp.getTime() < bucketTimestamp.getTime()) {
                        this.logger.error(`Record timestamp is before the smallest mutable bucket timestamp. Are your records out of order? Ignoring record '${record.data.id.value}'.`);
                        return [];
                    }
                    else if (this.members.length > 0 &&
                        recordTimestamp.getTime() <
                            this.members[this.members.length - 1].timestamp) {
                        this.logger.error(`Record timestamp is before the last record timestamp. Are your records out of order? Ignoring record '${record.data.id.value}'.`);
                        return [];
                    }
                    else if (recordTimestamp.getTime() >=
                        bucketTimestamp.getTime() + bucketSpan) {
                        bucket.immutable = true;
                        this.logger.debug(`Record timestamp is after the current bucket span. Making bucket '${bucket.id.value}' immutable.`);
                        this.mutableLeafBucketKeys.shift();
                        this.members = [];
                        bucketKey = this.mutableLeafBucketKeys[0];
                        this.makeParentImmutableIfNoMutableChildren(bucket, getBucket);
                    }
                    else {
                        candidateBucket = bucket;
                    }
                }
                while (this.members.length >= this.maxSize) {
                    const bucketNames = candidateBucket.id.value.split("/");
                    const bucketProperties = decodeURIComponent(bucketNames[bucketNames.length - 2]).split("_");
                    if (parseInt(bucketProperties[1]) / this.k <
                        this.minBucketSpan) {
                        this.logger.debug(`We need to make a new page. Bucket '${candidateBucket.id.value}' is full.`);
                        const newBucket = getBucket(`${bucketProperties[0]}_${bucketProperties[1]}_${parseInt(bucketProperties[2]) + 1}`);
                        candidateBucket.immutable = true;
                        addRelation(candidateBucket, newBucket, TREE.terms.GreaterThanOrEqualToRelation, literal(timestamp, namedNode(XSD.dateTime)), this.pathQuads);
                        this.members = [];
                        this.mutableLeafBucketKeys[0] = newBucket.id.value;
                        candidateBucket = newBucket;
                    }
                    else {
                        this.logger.debug(`We need to split the bucket '${candidateBucket.id.value}'.`);
                        const newBucketSpan = Math.round(parseInt(bucketProperties[1]) / this.k);
                        const newMutableLeafBuckets = [];
                        const parentBucket = candidateBucket;
                        parentBucket.empty = true;
                        parentBucket.immutable = true;
                        const recordTime = new Date(timestamp).getTime();
                        for (let i = 0; i < this.k; i++) {
                            const newTimestamp = new Date(new Date(bucketProperties[0]).getTime() +
                                i * newBucketSpan);
                            const newBucket = getBucket(`${newTimestamp.toISOString()}_${newBucketSpan}_0`);
                            this.logger.debug(`Creating new bucket '${newBucket.id.value}' with timespan ${newBucketSpan}.`);
                            this.members
                                .filter((m) => m.timestamp >= newTimestamp.getTime() &&
                                m.timestamp <
                                    newTimestamp.getTime() +
                                        newBucketSpan)
                                .forEach((m) => newBucket.addMember(m.id));
                            if (recordTime <
                                newTimestamp.getTime() + newBucketSpan) {
                                newMutableLeafBuckets.push(newBucket.id.value);
                                if (recordTime >= newTimestamp.getTime()) {
                                    this.members = this.members.filter((m) => m.timestamp >=
                                        newTimestamp.getTime() &&
                                        m.timestamp <
                                            newTimestamp.getTime() +
                                                newBucketSpan);
                                    if (this.members.length > 0) {
                                        const lastTimestamp = this.members[this.members.length - 1].timestamp;
                                        if (recordTime < lastTimestamp) {
                                            this.logger.error(`Record timestamp is before the last record timestamp in the new split bucket. Are your records out of order? Ignoring record '${record.data.id.value}'.`);
                                            return [];
                                        }
                                    }
                                    candidateBucket = newBucket;
                                }
                            }
                            else {
                                newBucket.immutable = true;
                                this.logger.debug(`Making bucket '${newBucket.id.value}' immutable.`);
                            }
                            this.addTimestampRelations(parentBucket, newBucket, newTimestamp, newBucketSpan, addRelation);
                        }
                        this.mutableLeafBucketKeys = [
                            ...newMutableLeafBuckets,
                            ...this.mutableLeafBucketKeys.slice(1),
                        ];
                    }
                }
                if (this.members.length > 0) {
                    const lastTimestamp = this.members[this.members.length - 1].timestamp;
                    if (new Date(timestamp).getTime() < lastTimestamp) {
                        this.logger.error(`Record timestamp is before the last record timestamp. Are your records out of order? Ignoring record '${record.data.id.value}'.`);
                        return [];
                    }
                }
                this.members.push({
                    id: JSON.parse(JSON.stringify(record.data.id.value)),
                    timestamp: new Date(timestamp).getTime(),
                });
                out.push(candidateBucket);
                this.logger.debug(`Added record '${record.data.id.value}' to bucket '${candidateBucket.id.value}'.`);
            }
            else {
                this.logger.warn(`Received records without timestamp values. Ignoring record '${record.data.id.value}'.`);
            }
        }
        return out;
    }
    save() {
        return JSON.stringify({
            mutableLeafBucketKeys: this.mutableLeafBucketKeys,
            members: this.members,
        });
    }
    addTimestampRelations(rootBucket, childBucket, startTimestamp, timespan, addRelation) {
        addRelation(rootBucket, childBucket, TREE.terms.GreaterThanOrEqualToRelation, literal(startTimestamp.toISOString(), namedNode(XSD.dateTime)), this.pathQuads);
        addRelation(rootBucket, childBucket, TREE.terms.LessThanRelation, literal(new Date(startTimestamp.getTime() + timespan).toISOString(), namedNode(XSD.dateTime)), this.pathQuads);
    }
    makeParentImmutableIfNoMutableChildren(bucket, getBucket) {
        const parent = bucket.parent;
        if (parent && !parent.root) {
            const children = parent.links.map((link) => {
                return getBucket(link.target.value, false, true);
            });
            const mutableChild = children.find((child) => !child.immutable);
            if (mutableChild === undefined) {
                parent.immutable = true;
                this.logger.debug(`Making parent bucket '${parent.id.value}' immutable.`);
                this.makeParentImmutableIfNoMutableChildren(parent, getBucket);
            }
        }
    }
}
