// @rdfc/sds-processors-ts/lib/bucketizers.js
import { readFileSync, writeFileSync } from "fs";
import { Parser, Writer as N3Writer } from "n3";
import { DataFactory } from "rdf-data-factory";
import { getLatestShape, getLatestStream, transformMetadata } from "./core.js";
import { LDES, PPLAN, PROV, RDF, SDS, XSD } from "@treecg/types";
import { BucketizerOrchestrator } from "./bucketizers/index.js";
import { Extractor, Record } from "./utils/index.js";
import { CBDShapeExtractor } from "extract-cbd-shape";
import { handleExit } from "./exitHandler.js";
import { RdfStore } from "rdf-stores";
const df = new DataFactory();
async function writeState(path, content) {
    if (path) {
        writeFileSync(path, content, { encoding: "utf-8" });
    }
}
function addProcess(id, store, strategyId, bucketizeConfig) {
    const newId = df.blankNode();
    const time = new Date().toISOString();
    store.addQuad(df.quad(newId, RDF.terms.type, PPLAN.terms.Activity));
    store.addQuad(df.quad(newId, RDF.terms.type, LDES.terms.Bucketization));
    bucketizeConfig.forEach((q) => store.addQuad(q));
    store.addQuad(df.quad(newId, PROV.terms.startedAtTime, df.literal(time)));
    store.addQuad(df.quad(newId, PROV.terms.used, strategyId));
    if (id) {
        store.addQuad(df.quad(newId, PROV.terms.used, id));
    }
    return newId;
}
function parseQuads(quads) {
    if (quads instanceof Array)
        return quads;
    const parser = new Parser();
    return parser.parse(quads);
}
function serializeQuads(quads) {
    const writer = new N3Writer();
    return writer.quadsToString(quads);
}
export function record_to_quads(record, resultingStream, buckets) {
    const id = df.blankNode();
    const out = [
        df.quad(id, SDS.terms.payload, record.data.id, SDS.terms.custom("DataDescription")),
        df.quad(id, SDS.terms.stream, resultingStream, SDS.terms.custom("DataDescription")),
        ...buckets
            .map((bucket) => bucket.id)
            .map((bucket) => df.quad(id, SDS.terms.bucket, bucket, SDS.terms.custom("DataDescription"))),
        ...(record.data.quads.length
            ? record.data.quads
            : [
                df.quad(id, SDS.terms.custom("dataless"), df.literal("true", XSD.terms.custom("boolean")), SDS.terms.custom("DataDescription")),
            ]),
    ];
    return out;
}
function bucket_to_quads(bucket, stream) {
    const out = [
        df.quad(bucket.id, RDF.terms.type, SDS.terms.custom("Bucket"), SDS.terms.custom("DataDescription")),
    ];
    out.push(df.quad(bucket.id, SDS.terms.custom("immutable"), df.literal((bucket.immutable || false) + ""), SDS.terms.custom("DataDescription")));
    if (bucket.root) {
        out.push(df.quad(bucket.id, SDS.terms.custom("isRoot"), df.literal("true"), SDS.terms.custom("DataDescription")));
    }
    if (bucket.empty) {
        out.push(df.quad(bucket.id, SDS.terms.custom("empty"), df.literal("true"), SDS.terms.custom("DataDescription")));
        bucket.empty = false;
    }
    if (stream) {
        out.push(df.quad(bucket.id, SDS.terms.stream, stream, SDS.terms.custom("DataDescription")));
    }
    return out;
}
function relationToQuads(bucket, rel, stream, remove) {
    const graph = remove
        ? SDS.terms.custom("RemoveDataDescription")
        : SDS.terms.custom("DataDescription");
    const out = [];
    const id = df.blankNode();
    out.push(df.quad(bucket.id, SDS.terms.relation, id, graph), df.quad(id, RDF.terms.type, SDS.terms.custom("Relation"), graph), df.quad(id, SDS.terms.relationType, rel.type, graph), df.quad(id, SDS.terms.relationBucket, rel.target, graph));
    if (rel.path) {
        out.push(df.quad(id, SDS.terms.relationPath, rel.path.id, graph), ...rel.path.quads.map((x) => df.quad(x.subject, x.predicate, x.object, graph)));
    }
    if (rel.value) {
        out.push(df.quad(id, SDS.terms.relationValue, rel.value, graph));
    }
    if (stream) {
        out.push(df.quad(bucket.id, SDS.terms.stream, stream, graph));
    }
    return out;
}
function set_metadata(channels, resultingStream, sourceStream, config) {
    const f = transformMetadata(resultingStream, sourceStream, "https://w3id.org/sds#Member", (x, y) => addProcess(x, y, config.quads.id, config.quads.quads));
    channels.metadataInput.data(async (quads) => channels.metadataOutput.push(serializeQuads(await f(parseQuads(quads)))));
    channels.metadataInput.on("end", async () => {
        await channels.metadataOutput.end();
    });
}
function read_save(savePath) {
    try {
        if (savePath) {
            return readFileSync(savePath, { encoding: "utf8" });
        }
    }
    catch (ex) {
        return;
    }
}
export async function bucketize(channels, config, savePath, sourceStream, resultingStream, prefix = "root") {
    set_metadata(channels, resultingStream, sourceStream, config);
    const save = read_save(savePath);
    const orchestrator = new BucketizerOrchestrator(config.strategy, save);
    const extractor = new Extractor(new CBDShapeExtractor(), sourceStream);
    channels.metadataInput.data(async (x) => {
        const quads = new Parser().parse(x);
        const store = RdfStore.createDefault();
        quads.forEach((q) => store.addQuad(q));
        const latest = sourceStream || (await getLatestStream(store));
        const latestShape = latest
            ? await getLatestShape(latest, store)
            : undefined;
        if (latestShape) {
            const rdfStore = RdfStore.createDefault();
            quads.forEach((x) => rdfStore.addQuad(x));
            const cbd_extract = new CBDShapeExtractor(rdfStore);
            extractor.extractor = cbd_extract;
            extractor.shape = latestShape;
        }
    });
    handleExit(async () => {
        const state = orchestrator.save();
        await writeState(savePath, state);
    });
    const buckets = {};
    channels.dataInput.data(async (x) => {
        const outputQuads = [];
        const quads = new Parser().parse(x);
        const records = await extractor.parse_records(quads);
        const requestedBuckets = new Set();
        const newMembers = new Map();
        const newRelations = [];
        const removeRelations = [];
        for (const record of records) {
            const record_buckets = orchestrator.bucketize(record, buckets, requestedBuckets, newMembers, newRelations, removeRelations, prefix);
            record_buckets.forEach((x) => requestedBuckets.add(x));
            outputQuads.push(...record_to_quads(record, resultingStream, record_buckets.map((x) => buckets[x])));
        }
        for (const [bucket, members] of newMembers) {
            if (buckets[bucket].empty) {
                continue;
            }
            for (const member of members) {
                outputQuads.push(...record_to_quads(new Record({ id: df.namedNode(member), quads: [] }, resultingStream, buckets[bucket]), resultingStream, [buckets[bucket]]));
            }
            requestedBuckets.add(bucket);
        }
        for (const requestedBucket of requestedBuckets) {
            outputQuads.push(...bucket_to_quads(buckets[requestedBucket], resultingStream));
        }
        for (const { origin, relation } of newRelations) {
            outputQuads.push(...relationToQuads(origin, relation));
        }
        for (const { origin, relation } of removeRelations) {
            outputQuads.push(...relationToQuads(origin, relation, resultingStream, true));
        }
        await channels.dataOutput.push(new N3Writer().quadsToString(outputQuads));
    });
    channels.dataInput.on("end", async () => {
        await writeState(savePath, orchestrator.save());
        await channels.dataOutput.end();
    });
}
