// @rdfc/sds-processors-ts/lib/sdsify.js
import { LDES, RDF, SDS, SHACL, XSD } from "@treecg/types";
import { DataFactory } from "rdf-data-factory";
import { getSubjects, maybeParse } from "./utils/index.js";
import { Writer as NWriter } from "n3";
import { CBDShapeExtractor } from "extract-cbd-shape";
import { RdfStore } from "rdf-stores";
import { createHash } from "crypto";
import { getLoggerFor } from "./utils/logUtil.js";
const logger = getLoggerFor("sdsify");
const df = new DataFactory();
async function extractMainNodeShape(store) {
    if (store) {
        const nodeShapes = await getSubjects(store, RDF.terms.type, SHACL.terms.NodeShape);
        let mainNodeShape = null;
        if (nodeShapes && nodeShapes.length > 0) {
            for (const ns of nodeShapes) {
                const referenced = await getSubjects(store, undefined, ns);
                const isNotReferenced = referenced.length === 0;
                if (isNotReferenced) {
                    if (!mainNodeShape) {
                        mainNodeShape = ns;
                    }
                    else {
                        throw new Error("There are multiple main node shapes in a given shape. " +
                            "Use sh:xone or sh:or to provide multiple unrelated shapes together.");
                    }
                }
            }
            if (mainNodeShape) {
                return mainNodeShape;
            }
            else {
                throw new Error("No main SHACL Node Shapes found in given shape filter");
            }
        }
        else {
            throw new Error("No SHACL Node Shapes found in given shape filter");
        }
    }
}
function getExtractor(shapeStore) {
    if (!shapeStore) {
        return new CBDShapeExtractor();
    }
    else {
        return new CBDShapeExtractor(shapeStore, undefined, {
            fetch: async () => new Response("", {
                headers: { "content-type": "text/turtle" },
            }),
        });
    }
}
export async function sdsify(input, output, streamNode, types, timestampPath, shape) {
    const shapeStore = shape ? RdfStore.createDefault() : null;
    if (shape) {
        maybeParse(shape).forEach((x) => {
            if (shapeStore)
                shapeStore.addQuad(x);
        });
    }
    const extractor = getExtractor(shapeStore);
    const shapeId = shape ? await extractMainNodeShape(shapeStore) : undefined;
    input.data(async (input) => {
        const dataStore = RdfStore.createDefault();
        maybeParse(input).forEach((x) => dataStore.addQuad(x));
        logger.debug(`Got input with ${dataStore.size} quads`);
        const members = {};
        const t0 = new Date();
        const subjects = [];
        if (types?.length) {
            for (const t of types) {
                subjects.push(...(await getSubjects(dataStore, RDF.terms.type, t)));
            }
        }
        else {
            subjects.push(...(await getSubjects(dataStore)));
        }
        await Promise.all(subjects.map(async (subject) => {
            if (subject.termType === "NamedNode" &&
                !members[subject.value]) {
                const membQuads = await extractor.extract(dataStore, subject, shapeId);
                members[subject.value] = {
                    quads: membQuads,
                    timestamp: timestampPath
                        ? dataStore.getQuads(subject, timestampPath)[0]
                            .object
                        : undefined,
                };
            }
        }));
        logger.debug(`Members extracted in ${new Date().getTime() - t0.getTime()} ms`);
        const orderedMembersIds = Object.keys(members);
        if (timestampPath) {
            orderedMembersIds.sort((a, b) => {
                const ta = new Date(members[a].timestamp.value).getTime();
                const tb = new Date(members[b].timestamp.value).getTime();
                return ta - tb;
            });
        }
        let membersCount = 0;
        const hash = createHash("md5");
        const TRANSACTION_ID = hash
            .update(new NWriter().quadsToString(dataStore.getQuads()))
            .digest("hex") +
            "_" +
            new Date().toISOString();
        for (const sub of orderedMembersIds) {
            const quads = members[sub].quads;
            const blank = df.blankNode();
            quads.push(df.quad(blank, SDS.terms.payload, df.namedNode(sub), SDS.terms.custom("DataDescription")), df.quad(blank, SDS.terms.stream, streamNode, SDS.terms.custom("DataDescription")), df.quad(blank, LDES.terms.custom("transactionId"), df.literal(TRANSACTION_ID), SDS.terms.custom("DataDescription")));
            if (membersCount === Object.keys(members).length - 1) {
                quads.push(df.quad(blank, LDES.terms.custom("isLastOfTransaction"), df.literal("true", XSD.terms.custom("boolean")), SDS.terms.custom("DataDescription")));
            }
            await output.push(new NWriter().quadsToString(quads));
            membersCount += 1;
        }
        logger.debug(`Successfully pushed ${membersCount} members in ${new Date().getTime() - t0.getTime()} ms`);
    });
    input.on("end", async () => {
        logger.info("Input channel was closed down");
        await output.end();
    });
}
